---
title: "deleted_code"
author: "Álvaro Gómez Pérez"
date: "2025-02-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```







# Old DE code before I made it into functions in helperfunctions package
### 1.2. Perform DE analysis - with limpa! Using this one from now on
After loading results from protein inference step - this step was run once and took extremely long, so we just saved the results and load them now
```{r}
# Prepare to perform the DEA
## Get the names of all KOs we have, and set WT as the reference
trna_levels <- c()
for (i in 1:ncol(trna_ko)) {
  sample_name <- colnames(trna_ko)[i]
  strain_name <- sample_layout$Strain.Name[sample_layout$final_proteomics_colnames == sample_name]
  if (length(strain_name) > 1) {
    strain_name <- strain_name[1]
  }
  trna_levels <- c(trna_levels, strain_name)
}
trna_levels <- as.factor(trna_levels)
trna_levels <- relevel(trna_levels, ref = "WT")

## Generate design matrix
mm <- model.matrix(~ 0 + trna_levels)
colnames(mm) <- levels(trna_levels)

fit <- dpcDE(y.protein, mm)
```

Perform the DE analysis - THIS IS INCLUDED IN MY FUNCTION
```{r}
# This is what they provide, I wonder if they use the kind of design matrix that Alexis used
# I did not, so I am going to introduce contrasts here? I hope that makes sense?
## Create all contrasts - need to susbtitute parentheses in strain names by dots so that it runs...
treatments <- colnames(fit$coefficients)
treatments <- sapply(treatments, function(x) str_replace(x, "\\(", "."))
treatments <- sapply(treatments, function(x) str_replace(x, "\\)", "."))
design <- model.matrix(~ 0 + treatments)
colnames(design) <- treatments
contrast_list <- paste0(treatments, "-WT")
contrast_list <- contrast_list[!(contrast_list %in% c("WT-WT"))]          # Remove the contrasts of the WT to itself

## Need to change parentheses to dots in the fit object as well so they match...
colnames(fit$coefficients) <- sapply(colnames(fit$coefficients), function(x) str_replace(x, "\\(", "."))
colnames(fit$coefficients) <- sapply(colnames(fit$coefficients), function(x) str_replace(x, "\\)", "."))

## Introduce the contrasts and run eBayes for them, 10 at a time
da <- list()
batch_size <- 10
n_batches <- ceiling(ncol(fit$coefficients)/batch_size)

for (i in 1:n_batches) {
  contrast_subset <- contrast_list[((i - 1) * batch_size + 1):min(i * batch_size, length(contrast_list))]
  contrast.matrix <- makeContrasts(contrasts = contrast_subset, levels = design)
  
  fit2 <- contrasts.fit(fit, contrast.matrix)
  fit3 <- eBayes(fit2, robust = TRUE, trend = TRUE)
  
  new_contrast_names <- c()
  
  # Store topTables for each contrast
  for (contrast_name in colnames(contrast.matrix)) {
    print(contrast_name)
    new_name <- substr(contrast_name, 1, str_locate(contrast_name, "-")-1)
    if (grepl(".", new_name)) {
        substr(new_name, 3, 3) <- "("
        substr(new_name, 7, 7) <- ")"
    }
    da[[new_name]] <- topTable(fit3, coef = contrast_name, adjust.method = 'BH', number = Inf, sort.by = 'none')
    da[[new_name]]$Strain.Name <- rep(new_name, nrow(da[[new_name]]))
  }
}
```

Save DE proteins for GO enrichment analysis - NOT YET IN MY FUNCTION - IS IT NECESSARY IF I AM SAVING ALL PROTEINS BELOW ALREADY AND CAN DO GSEA WITH GO TERMS? I DON'T THINK SO
```{r}
# Just a list of DE proteins per strain
yeastmine <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/databases/alliancemine_results_2024-11-20T11-19-04.tsv", sep="")))
de_proteins_list <- list()
for (i in 1:length(da)) {
  temp <- da[[i]]
  temp <- na.omit(temp)
  
  # Collect protein names and turn them all into systematic notation
  general_protein_names <- rownames(temp)[(temp$adj.P.Val < alpha) &
                                          (temp$logFC > lfc_threshold | temp$logFC < -lfc_threshold)]
  standard_protein_names <- general_protein_names[general_protein_names %in% yeastmine$Gene.symbol]
  systematic_protein_names <- general_protein_names[general_protein_names %in% yeastmine$Gene.secondaryIdentifier]
  final_protein_names <- c(systematic_protein_names, match_systematic_and_standard_protein_names(data = standard_protein_names,
                                                                                                 uniprot = uniprot_db,
                                                                                                 input = "standard",
                                                                                                 output = "systematic",
                                                                                                 simplify = T))
  
  # Save protein names
  de_proteins_list[[names(da)[i]]] <- final_protein_names
}
output_file <- paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/de_proteins_list_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep="")
write_json(de_proteins_list, path=output_file)
```

Save all proteins for GSEA - THIS IS INCLUDED IN MY FUNCTION
```{r}
#write_json(da, path=paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/da_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep=""), pretty = T)
# MOVED THIS HALFWAY THROUGH THE NEXT CHUNK SO IT WOULD CONTAIN UP- OR DOWN-REGULATED INFORMATION - HOPE THIS DOESN'T BREAK ANYTHING - 03.06.2025
```

Finish processing the DE analysis results - THIS IS INCLUDED IN MY FUNCTION - JUST NEED TO ADD THE STEPS WHERE I SAVE THINGS
```{r}
# Format data
da <- bind_rows(da) %>%
  dplyr::mutate(protein = rep(rownames(da[[1]]), times = length(da))) %>%
  dplyr::distinct(Strain.Name, protein, logFC, .keep_all = T)
temp <- sample_layout %>%
  dplyr::distinct(Strain.Name, .keep_all = T)
da <- left_join(da, temp, by = "Strain.Name") %>%
  dplyr::relocate(Strain.Name, .before = logFC) %>%
  dplyr::relocate(protein, .after = Strain.Name) %>%
  filter(Strain.Name != "WT")

da <- da %>%
  dplyr::select(protein, Strain.Name, logFC, P.Value, adj.P.Val) %>%                              # From here on in this function it's added by me
  dplyr::mutate(diffexpressed_adjusted = case_when((logFC > lfc_threshold) & (adj.P.Val < alpha) ~ "Up_regulated",
                                            (logFC < -lfc_threshold) & (adj.P.Val < alpha) ~ "Down_regulated",
                                            TRUE ~ "Not_significant"),
         diffexpressed_non_adjusted = case_when((logFC > lfc_threshold) & (P.Value < alpha) ~ "Up_regulated",
                                                (logFC < -lfc_threshold) & (P.Value < alpha) ~ "Down_regulated",
                                                TRUE ~ "Not_significant"))

da <- da %>%
  dplyr::distinct(Strain.Name, protein, logFC, .keep_all = T)

write_json(da, path=paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/da_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep=""), pretty = T)


# Add a column to da with the number of replicates per KO - by me
unique_KOs <- unique(da$Strain.Name)
replicates <- c()
for (i in 1:length(unique_KOs)) {
  KO <- unique_KOs[i]
  replicates <- c(replicates, sum(grepl(KO, colnames(trna_ko), fixed = T)))
}
KOs_replicates <- data.frame(unique_KOs, replicates)
colnames(KOs_replicates) <- c("Strain.Name", "Replicate_num")
da <- merge(da, KOs_replicates, by = "Strain.Name")

# Responsiveness
responsiveness <- da %>%
  group_by(Strain.Name) %>%
  summarise(nDEP = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha, na.rm = T),
            Up_regulated_adjusted = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha & diffexpressed_adjusted == "Up_regulated", na.rm = T),
            Down_regulated_adjusted = sum(abs(logFC) >= lfc_threshold & adj.P.Val <= alpha & diffexpressed_adjusted == "Down_regulated", na.rm = T),
            Up_regulated_non_adjusted = sum(abs(logFC) >= lfc_threshold & P.Value <= alpha & diffexpressed_non_adjusted == "Up_regulated", na.rm = T),
            Down_regulated_non_adjusted = sum(abs(logFC) >= lfc_threshold & P.Value <= alpha & diffexpressed_non_adjusted == "Down_regulated", na.rm = T),
            Replicate_num = mean(Replicate_num)) %>%
  mutate(Amino_acid_1_letter = str_sub(Strain.Name, 2, 2),
         chromosome_letter = substr(Strain.Name, 8, 8),
         anticodon = str_extract(Strain.Name, "(?<=\\()[[:alpha:]]{3}(?=\\))"),
         up_down_regulated_ratio_adjusted = Up_regulated_adjusted/Down_regulated_adjusted,
         up_down_regulated_ratio_non_adjusted = Up_regulated_non_adjusted/Down_regulated_non_adjusted) %>%
  arrange(Amino_acid_1_letter) %>%
  mutate(anticodon = factor(anticodon, levels = unique(anticodon)))

# Add all amino acid names
amino_acids <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/databases/GtRNAdb/amino_acids.csv", sep="")))
responsiveness <- left_join(responsiveness, amino_acids, by = "Amino_acid_1_letter")

# Save responsiveness dataset
fwrite(responsiveness, paste(base_dir, "tRNA_KOs/Data/produced_during_analysis/responsiveness_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".csv", sep=""))

# Save results from the DE analysis in the same way as in the file where I do DE separately for each batch
DE <- list(fit = fit,
            fit2 = fit2,
            da = da, 
            responsiveness= responsiveness)



# Remove unnecesary variables
rm(amino_acids, DE, de_proteins_list, fit, fit2, KOs_replicates, mm, temp, trna_levels, contrast.matrix, design, fit3,
   yeastmine, final_protein_names, general_protein_names, i, KO, output_file, replicates, sample_name, standard_protein_names, strain_name,
   systematic_protein_names, unique_KOs, trna_ko, dpcfit)
```

Save DE proteins (up- and down-regulated separately) - THIS IS INCLUDED IN MY FUNCTION
```{r}
de_proteins_up_down <- list()
strains <- unique(da$Strain.Name)

for (i in 1:length(strains)) {
  strain <- strains[i]
  temp <- da %>%
    dplyr::filter(Strain.Name == strain)
  
  up_regulated <- temp$protein[temp$diffexpressed_adjusted == "Up_regulated"]
  down_regulated <- temp$protein[temp$diffexpressed_adjusted == "Down_regulated"]
  
  de_proteins_up_down[[strain]] <- list("Up" = up_regulated,
                                        "Down" = down_regulated)
}


# Save
output_file <- paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/de_proteins_list_up_down_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep="")
write_json(de_proteins_up_down, path=paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/de_proteins_list_up_down_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep=""))
```






















# 3. from tRNAs_together.Rmd

Compare to expression data from Cozen et al., 2015 - THESE ARE JUST READ COUNTS I'M LOOKING AT, HONESTLY NO IDEA IF THIS MAKES SENSE
The main issue with this dataset (apart from the fact that they are read counts) is that in many rows, the counts are not assigned to a single tRNA gene, but they can come from any tRNA gene with a certain anticodon. Not sure how to go about this, so for now I am going to try to look only at the rows which identify one single tRNA gene, and see if that is enough for the analysis

I think that maybe the different rows that correspond to the same tRNA (including those rows that map to different tRNAs) occur because each of these rows comes from reads mapping to a certain section of that tRNA molecule??? In that case, it makes much more sense to sum across rows for the same tRNA, instead of taking the mean. In fact, I could sum those which are mapped to different tRNAs as well. I should do that. 

--------> I think this section is now pretty useless, the correct one is 4., where I do this only using genomic tRNA sequencing counts


## 3.1. Come up with plots with nDEP vs. expression - using all rows

Split rows that are mapped to multiple tRNAs 
```{r}
cozen <- read.xlsx(paste(base_dir, "Data/Other/tRNA_expression_comparative/cozen_et_al_2015.xlsx", sep=""), 1)

# Split rows with multiple genes identified into multiple rows with a single gene, with the same values in all other columns - add a column identifying whether each row came from an original row with multiple tRNAs or not
cozen <- cozen %>% 
  dplyr::select(-c(corresponding.mature.tRNA.or.tRNA.locus.names..RNA.central., genomic.tRNA.database.legacy.names))

new_column <- c()
cozen_split <- data.frame(matrix(nrow = 0, ncol = ncol(cozen)))
colnames(cozen_split) <- colnames(cozen)

for (i in 1:nrow(cozen)) {
  trna_names <- cozen$saccaromyces.genome.datasbase..sgd..tRNA.name[i]
  
  if (!grepl(",", trna_names)) {
    cozen_split <- rbind(cozen_split, cozen[i,])
    new_column <- c(new_column, "Single")
  }
  
  else {
    trnas_mapped_to <- strsplit(trna_names, ",")[[1]]
    for (j in 1:length(trnas_mapped_to)) {
      new_row <- cozen[i,]
      new_row[which(colnames(cozen) == "saccaromyces.genome.datasbase..sgd..tRNA.name")] <- trnas_mapped_to[j]
      cozen_split <- rbind(cozen_split, new_row)
      new_column <- c(new_column, "Multiple")
    }
  }
}
cozen_split$Num_of_tRNAs_originally_in_row <- new_column
```

Old code where I simply got rid of those rows - NOT RUN NOW
```{r}
## Keep only rows with a single tRNA gene identified, remove those with unknown (NA) gene identified and with NA in the mean.normalized.ReadCount..AlkB column
#cozen_single_to_merge <- cozen %>% 
#  filter(!grepl(",", saccaromyces.genome.datasbase..sgd..tRNA.name) & 
#           saccaromyces.genome.datasbase..sgd..tRNA.name != "NA" &
#           read.mapping..genomic.sequence.or.mature.tRNA. == "genomic" &
#           mean.normalized.ReadCount..AlkB != "NA") %>%
#  dplyr::select(name, tRNA.group, chromosome, chromosome.start, chromosome.end, chromosome.start, saccaromyces.genome.datasbase..sgd..tRNA.name, mean.normalized.ReadCount..AlkB) %>%
#  group_by(saccaromyces.genome.datasbase..sgd..tRNA.name) %>%
#  mutate(mean_reads_per_tRNA = sum(as.numeric(mean.normalized.ReadCount..AlkB), na.rm = T)) %>%
#  rename(Strain.Name = saccaromyces.genome.datasbase..sgd..tRNA.name)
```

Remove rows with NAs in relevant columns, select columns of interest, get the sum of reads per tRNA (since there are several rows for the same tRNA), and rename to merge with responsiveness dataframe
```{r}
cozen_to_merge <- cozen_split %>%
  filter(saccaromyces.genome.datasbase..sgd..tRNA.name != "NA" &
           mean.normalized.ReadCount..AlkB != "NA") %>%
  dplyr::select(name, tRNA.group, chromosome, chromosome.start, chromosome.end, chromosome.start, Num_of_tRNAs_originally_in_row,
         saccaromyces.genome.datasbase..sgd..tRNA.name, mean.normalized.ReadCount..AlkB, read.mapping..genomic.sequence.or.mature.tRNA.) %>%
  group_by(saccaromyces.genome.datasbase..sgd..tRNA.name) %>%
  mutate(mean_reads_per_tRNA = sum(as.numeric(mean.normalized.ReadCount..AlkB), na.rm = T)) %>%
  rename(gene_name = saccaromyces.genome.datasbase..sgd..tRNA.name)

## Merge with resp
resp_and_expression <- left_join(resp, cozen_to_merge, by = "gene_name")
```

Get plots
```{r}
# Plot average nDEP vs. avg expression - genomic tRNA sequences
ggplot(data = subset(resp_and_expression, read.mapping..genomic.sequence.or.mature.tRNA. == "genomic"), 
       aes(x = mean_reads_per_tRNA, y = nDEP, col = Num_of_tRNAs_originally_in_row)) +
  geom_point() +
  theme_light()

# Plot average nDEP vs. avg expression - mature tRNA sequences
ggplot(data = subset(resp_and_expression, read.mapping..genomic.sequence.or.mature.tRNA. == "mature_tRNA"), 
       aes(x = mean_reads_per_tRNA, y = nDEP, col = Num_of_tRNAs_originally_in_row)) +
  geom_point() +
  theme_light()
```
Would like to have more data points in the second one, I guess mature tRNAs are harder to sequence???


## 3.2. Integrate data from tRNA synthetase abundance and tRNA abundance (Chu et al., 2011) 
```{r}
# Load data
chu_synths <- as.data.frame(fread(paste(base_dir, "Data/Other/Articles/chu_2011/S3.csv", sep="")))
chu_trnas <- as.data.frame(fread(paste(base_dir, "Data/Other/Articles/chu_2011/S2.csv", sep="")))
# The problem with these abundances for tRNAs is that they are for each family, not for each tRNA concretely, so I don't really know if they will be useful to me. I could collapse those from Cozen et al. by family and compare them to this to see if they point in the same direction, but might be a bit of a time waste

# Join the synthetase data to the resp_and_expression dataset
amino_acids <- as.data.frame(fread(paste(base_dir, "Data/Other/GtRNAdb/amino_acids.csv", sep=""))) %>%
  rename(Amino_acid = X3_letter_code, 
         AA = X1_letter_code)

chu_synths <- left_join(chu_synths, amino_acids, by = "Amino_acid") %>%
  rename(synthetase_abundance = Abundance) %>%
  dplyr::select(-c(k_cat_value, k_cat_organism))
resp_and_expression <- left_join(resp_and_expression, chu_synths, by = "AA")
```

Plots
```{r}
# Plot tRNA abundance from Cozen et al., against tRNA synthetase abundance from Chu et al.
## Genomic tRNAs
ggplot(data = subset(resp_and_expression, read.mapping..genomic.sequence.or.mature.tRNA. == "genomic"), 
       aes(x = synthetase_abundance, y = mean_reads_per_tRNA)) +
  geom_point() +
  theme_light()

## Mature tRNAs
ggplot(data = subset(resp_and_expression, read.mapping..genomic.sequence.or.mature.tRNA. == "mature_tRNA"), 
       aes(x = synthetase_abundance, y = mean_reads_per_tRNA)) +
  geom_point() +
  theme_light()



# Same plots as before, tRNA abundance from Cozen et al. vs. our nDEP, coloring by tRNA synthetase abundance from Chu et al.
## Genomic tRNA sequences
ggplot(data = subset(resp_and_expression, read.mapping..genomic.sequence.or.mature.tRNA. == "genomic"), 
       aes(x = mean_reads_per_tRNA, y = nDEP, col = synthetase_abundance)) +
  geom_point() +
  theme_light()

## Mature tRNA sequences
ggplot(data = subset(resp_and_expression, read.mapping..genomic.sequence.or.mature.tRNA. == "mature_tRNA"), 
       aes(x = mean_reads_per_tRNA, y = nDEP, col = synthetase_abundance)) +
  geom_point() +
  theme_light()
```


## 3.3. Explore the "genomic" and "mature tRNAs" rows, they seem to be the same?
```{r}
genomic <- resp_and_expression %>%
  filter(read.mapping..genomic.sequence.or.mature.tRNA. == "genomic")

mature <- resp_and_expression %>%
  filter(read.mapping..genomic.sequence.or.mature.tRNA. == "mature_tRNA")
```












######################################################################################
# 6. Check if different KOs with the same anticodon cause the same proteins to be DE
######################################################################################
(From main tRNAs_together.Rmd - put here on 07.05.2024)


## 6.2. Upset plots
Example - to be deleted
```{r}
library(ggplot2movies)
movies = as.data.frame(ggplot2movies::movies)
head(movies, 3)

genres = colnames(movies)[18:24]
genres

movies[genres] = movies[genres] == 1
t(head(movies[genres], 3))

movies[movies$mpaa == '', 'mpaa'] = NA
movies = na.omit(movies)

upset(movies, genres, name='genre', width_ratio=0.1)
```

Per anticodon
```{r}
# Preparations: 
#   - Remove KOs with 0 DE proteins from list, 
#   - Grab anticodon names
#   - Create empty list to put data matrices for Upset plots there
de_protein_list <- Filter(function(x) length(x) > 0, de_proteins_list_001)
anticodons <- unique(responsiveness$anticodon)


#pdf(file = paste(base_dir, "Output/Plots/compare_DEPs_across_KOs/upset/upset_plots_per_anticodon.pdf", sep = ""), onefile = T, width = 12, height = 5, paper = 'USr')

# Iterate over the anticodons and get an Upset plot for each of them
for (i in 1:length(anticodons)) {
  # For each anticodon, grab the tRNAs that contain it, and get a matrix with the right format for an upset plot
  anticodon <- anticodons[i]
  KOs_with_anticodon <- de_protein_list[grepl(anticodon, names(de_protein_list))]
  KOs_with_anticodon_names <- names(KOs_with_anticodon)
  
  # This is so I don't process anticodons for which there is only one KO - there is an error if I do so
  if (length(KOs_with_anticodon) > 1) {
    # Create the matrix where I will add one-hot encoded columns for each KO
    all_proteins_anticodon <- unique(unlist(KOs_with_anticodon))
    df_to_upset <- data.frame(Proteins = all_proteins_anticodon)
    
    # Iterate over the KO strains and add a one-hot encoded column for each of them, for presence/absence of that protein among the DE prots in that strain
    for (j in 1:length(KOs_with_anticodon_names)) {
      KO <- KOs_with_anticodon_names[j]
      proteins <- KOs_with_anticodon[[KO]]
      df_to_upset[,KO] <- all_proteins_anticodon %in% proteins
    }
    
    # Now that we have the dataframe, get the Upset plot and save it to the list
    ## Using ComplexUpset
    ComplexUpset::upset(data = df_to_upset,
                        intersect = KOs_with_anticodon_names,
                        name='KO strain', 
                        width_ratio=0.1,
                        mode = "intersect") +
      ggtitle(label = "Intersection in DE proteins",
              subtitle = glue("Across KOs for tRNAs with {anticodon}"))
    
    
    image_name <- paste(base_dir, "Output/Plots/compare_DEPs_across_KOs/upset/per_anticodon/upset_", anticodon, ".png", sep="")
    ggsave(filename = image_name, width = 8, height = 5)
  }
}

#dev.off()
```

Per amino acid - TO DO
```{r}
# Preparations: 
#   - Remove KOs with 0 DE proteins from list, 
#   - Grab anticodon names
#   - Create empty list to put data matrices for Upset plots there
de_protein_list <- Filter(function(x) length(x) > 0, de_proteins_list_001)
aas <- unique(responsiveness$Amino_acid_1_letter)


pdf(file = paste(base_dir, "Output/Plots/compare_DEPs_across_KOs/upset/upset_plots_per_amino_acid.pdf", sep = ""), onefile = T, width = 12, height = 5, paper = 'USr')

# Iterate over the anticodons and get an Upset plot for each of them
for (i in 1:length(aas)) {
  # For each anticodon, grab the tRNAs that contain it, and get a matrix with the right format for an upset plot
  aa <- aas[i]
  KOs_with_anticodon <- de_protein_list[grepl(anticodon, names(de_protein_list))]
  KOs_with_anticodon_names <- names(KOs_with_anticodon)
  
  # This is so I don't process anticodons for which there is only one KO - there is an error if I do so
  if (length(KOs_with_anticodon) > 1) {
    # Create the matrix where I will add one-hot encoded columns for each KO
    all_proteins_anticodon <- unique(unlist(KOs_with_anticodon))
    df_to_upset <- data.frame(Proteins = all_proteins_anticodon)
    
    # Iterate over the KO strains and add a one-hot encoded column for each of them, for presence/absence of that protein among the DE prots in that strain
    for (j in 1:length(KOs_with_anticodon_names)) {
      KO <- KOs_with_anticodon_names[j]
      proteins <- KOs_with_anticodon[[KO]]
      df_to_upset[,KO] <- all_proteins_anticodon %in% proteins
    }
    
    # Now that we have the dataframe, get the Upset plot and save it to the list
    ## Using ComplexUpset
    p <- ComplexUpset::upset(data = df_to_upset,
                             intersect = KOs_with_anticodon_names,
                             name='KO strain', 
                             width_ratio=0.1,
                             mode = "intersect")
    print(p)
  }
}

dev.off()
```



### 6.3.2. Exploratory plots - HONESTLY I FEEL LIKE MOST OF THESE HISTOGRAMS ARE PRETTY USELESS, AREN'T THEY? COULD MAYBE DELETE THIS OR PUT IT IN THE DELETED CODE DOCUMENT?
Load dataset so I don't have to create it from scratch
```{r}
# Protein overlaps
if (alpha == 0.01) {
  DE_protein_overlaps <- as.data.frame(fread(paste(base_dir, "Data/Other/compare_DEPs_across_KOs/DE_protein_overlaps_001.tsv", sep="")))
}
if (alpha == 0.05) {
  DE_protein_overlaps <- as.data.frame(fread(paste(base_dir, "Data/Other/compare_DEPs_across_KOs/DE_protein_overlaps_005.tsv", sep="")))
}

# Master dataset
master_dataset <- as.data.frame(fread(paste(base_dir, "Data/Other/GtRNAdb/master_tRNA_dataset.csv", sep = "")))
```

Want to filter out some observations before the plots?
```{r}
#DE_protein_overlaps <- DE_protein_overlaps %>%
#  filter(protein_number_1 < 1000,
#         protein_number_2 < 1000)
```

Plots - histograms - color by family, amino acid, chromosome - THESE ONES I AM USING 
```{r}
ggplot(data = DE_protein_overlaps, aes(x = Jaccard_index)) +
  geom_histogram(alpha = 0.7, binwidth = 0.01) +
  theme_light() +
  xlab("Jaccard index") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains")

ggplot(data = subset(DE_protein_overlaps, Jaccard_index > 0), aes(x = Jaccard_index)) +
  geom_histogram(alpha = 0.7, binwidth = 0.01) +
  theme_light() +
  xlab("Jaccard index") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins - after removing all 0 values",
       subtitle = "For all possible pairs of KO strains") 

ggplot(data = subset(DE_protein_overlaps, Jaccard_index > 0), aes(x = Jaccard_index, fill = same_amino_acid)) +
  geom_histogram(alpha = 0.7, binwidth = 0.01) +
  theme_light() +
  xlab("Jaccard index") +
  ylab("Count") +
  labs(fill = "Did the 2 KOs carry\nthe same amino acid?") 

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index))) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins  - after removing all 0 values",
       subtitle = "For all possible pairs of KO strains") 

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = same_family)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "Same family?") +
  scale_fill_viridis_d()

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = same_amino_acid)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "Same amino acid?") +
  scale_fill_viridis_d()

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = same_chromosome)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "Same chromosome?") +
  scale_fill_viridis_d()

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = same_family_and_chr)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "Same family and\nchromosome?") +
  scale_fill_viridis_d()

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = same_amino_acid_and_chr)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "Same amino acid and\nchromosome?") +
  scale_fill_viridis_d()
```

Plots - histograms - by both aa charges and both aa types
```{r}
ggplot(data = DE_protein_overlaps, aes(x = both_aa_types, y = log2(Jaccard_index), fill = both_aa_types)) +
  geom_boxplot() +
  theme_light() +
  xlab("Amino acid types") +
  ylab("log2(Jaccard index)") +
  labs(title = "Boxplots of the Jaccard Index values for all pairs of KO strains",
       subtitle = "Classified by the type of the amino acids carried by the KOd tRNAs") +
  scale_fill_viridis_d() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))


ggplot(data = DE_protein_overlaps, aes(x = both_aa_charges, y = log2(Jaccard_index), fill = both_aa_charges)) +
  geom_boxplot() +
  theme_light() +
  xlab("Amino acid charges") +
  ylab("log2(Jaccard index)") +
  labs(title = "Boxplots of the Jaccard Index values for all pairs of KO strains",
       subtitle = "Classified by the charges of the amino acids carried by the KOd tRNAs") +
  scale_fill_viridis_d() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```

Plots - histograms - try to explain the larger overlaps
```{r}
ggplot(data = subset(DE_protein_overlaps, overlap_size > 0), aes(x = Jaccard_index)) +
  geom_histogram(alpha = 0.7, binwidth = 0.01) +
  theme_light() +
  xlab("Jaccard_index") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard_index of the DE proteins",
       subtitle = "For all possible pairs of KO strains") 




ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = Strain_more_DEPs, group = Strain_more_DEPs)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "DEPs in strain\nwith more DEPs") +
  scale_fill_viridis_c()

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = Strain_less_DEPs, group = Strain_less_DEPs)) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "DEPs in strain\nwith less DEPs") +
  scale_fill_viridis_c()

ggplot(data = DE_protein_overlaps, aes(x = log2(Jaccard_index), fill = log2(Ratio_more_to_less_DEPs), group = log2(Ratio_more_to_less_DEPs))) +
  geom_histogram(alpha = 0.7) +
  theme_light() +
  xlab("log2(Jaccard index)") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = "For all possible pairs of KO strains",
       fill = "nDEP ratio -\nlarger/smaller") +
  scale_fill_viridis_c()
```

Plots - histograms - separate histogram for each anticodon - don't think this is too useful
```{r}
plot_list <- list()
anticodons <- unique(responsiveness$anticodon)
for (i in 1:length(anticodons)) {
  anticodon <- anticodons[i]
  temp <- DE_protein_overlaps %>%
    filter(anticodon_1 == anticodon & anticodon_2 == anticodon)
  
  plot_list[[i]] <- ggplot(data = temp, aes(x = Jaccard_index)) +
  geom_histogram(alpha = 0.7, binwidth = 1) +
  theme_light() +
  xlab("Jaccard index") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = glue("For strains with {anticodon} anticodon")) +
  scale_fill_viridis_d()
}


## List of ggplot objects has to be turned to list of grobs for grid.arrange to work
my_grobs <- lapply(plot_list, ggplotGrob)

## Print to PDF
output_file = paste(base_dir, "Output/Plots/compare_DEPs_across_KOs/histograms/Jaccard_index_per_family.pdf", sep = "")
ggsave(
   filename = output_file, 
   plot = marrangeGrob(my_grobs, nrow=4, ncol=4), 
   width = 15, height = 9
)
```

Plots - histograms - separate histogram for each amino acid - don't think this is too useful
```{r}
plot_list <- list()
aas <- unique(responsiveness$Amino_acid_1_letter)
for (i in 1:length(aas)) {
  aa <- aas[i]
  temp <- DE_protein_overlaps %>%
    filter(Amino_acid_1 == aa & Amino_acid_2 == aa)
  
  plot_list[[i]] <- ggplot(data = temp, aes(x = Jaccard_index)) +
  geom_histogram(alpha = 0.7, binwidth = 1) +
  theme_light() +
  xlab("Jaccard index") +
  ylab("Count") +
  labs(title = "Histogram of Jaccard index of the DE proteins",
       subtitle = glue("For strains carrying {aa}")) +
  scale_fill_viridis_d()
}


## List of ggplot objects has to be turned to list of grobs for grid.arrange to work
my_grobs <- lapply(plot_list, ggplotGrob)

## Print to PDF
output_file = paste(base_dir, "Output/Plots/compare_DEPs_across_KOs/histograms/Jaccard_index_per_amino_acid.pdf", sep = "")
ggsave(
   filename = output_file, 
   plot = marrangeGrob(my_grobs, nrow=4, ncol=4), 
   width = 15, height = 9
)
```




































```{r}
grab_de_synthetases_dfs <- function(da_full, de_synthetases, yeastmine) {
  de_synthetases_dfs <- list()
  for (i in 1:length(da_full)) {
    strain_name <- names(da_full)[[i]]
    strains_with_de_synths <- names(de_synthetases)
    if (strain_name %in% strains_with_de_synths) {
      print(i)
      temp <- da_full[[i]]
      rownames(temp) <- match_systematic_and_standard_protein_names(data = rownames(temp),
                                                                       yeastmine = yeastmine,
                                                                       input = "standard",
                                                                       simplify = T) 
      temp <- temp[rownames(temp) %in% de_synthetases[[strain_name]],]
      temp$Synthetase <- rownames(temp)
      temp$Strain.Name <- strain_name
      de_synthetases_dfs[[strain_name]] <- temp
    }
  }
  
  # Turn the resulting list into a single dataframe
  de_synthetases_df <- bind_rows(de_synthetases_dfs)
  rownames(de_synthetases_df) <- NULL
  
  # Add a column with the standard names for the tRNA synthetase genes
  temp <- match_systematic_and_standard_protein_names(data = as.character(de_synthetases_df$Synthetase),
                                                      yeastmine = yeastmine, 
                                                      input = "systematic",
                                                      simplify = T)
  de_synthetases_df$Synthetase_standard <- temp
  de_synthetases_df <- de_synthetases_df %>%
    rename(Synthetase_systematic = Synthetase)
  
  return(de_synthetases_dfs)
}


da <- read_json(path=paste(base_dir, "Data/Other/enrichment_analysis/da.json", sep=""))

de_synthetases_dfs_001 <- grab_de_synthetases_dfs(da, de_synthetases_001, yeastmine)
de_synthetases_dfs_005 <- grab_de_synthetases_dfs(da, de_synthetases_005, yeastmine)
```