---
title: "correlation_with_pilpel_2014"
author: "Álvaro Gómez Pérez"
date: "2024-11-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```




############################### DISCLAIMER ############################### 
This I just cut from the main file and pasted here quite some time ago, so it is not fixed to work at all. 
Today (20.03.2025) I have tried to make it work to obtain the original correlation plots but for the newly pre-processed data (batch correction per subbatch) - I have only made it to 2.1. (included) - SHOULD COME BACK SOME TIME AND FIX THE REST? BUT NOT MY MAIN CONCERN I GUESS? IDK
 
 
## Updated on 05.09.2025 - all working properly :)





Packages
```{r}
library(dplyr)
library(data.table)
library(limma)
library(ggplot2)
library(patchwork)
library(tidyr)
library(tibble)
library(stringr)
library(gprofiler2)
library(xlsx)
library(stringr)
library(gridExtra)
library(grid)
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(circlize))
library(factoextra)
library(glue)
library(dendextend)
library(ggrepel)
library(ggvenn)
library(tidyr)
library(interactions)
library(performance)
library(MASS)
library(jsonlite)
library(helperfunctions)
```

Set parameters
```{r}
# Significance level to be used for all tests and plots in this file
alpha <- 0.05

# Significance level as a plain string to use when loading or writing files
alpha_plain <- str_replace(as.character(alpha), "\\.", "")

# Minimum threshold for the average log2-expression across all samples in order to keep a protein in the data before DE analysis
mean_log2_across_all_samples_threshold <- 2

# Minimum threshold for the variance of the log2-expression across all samples in order to keep a protein in the data before DE analysis
var_across_log2_all_samples_threshold <- 0.6

# Separate significance level, the one used for the enrichment analysis
alpha_enrichment <- 0.05

# Significance level above as a plain string to use when loading or writing files
alpha_enrichment_plain <- str_replace(as.character(alpha_enrichment), "\\.", "")

# Log-fold change limit to be considered "biologically significant"
lfc_threshold <- 0.5

# Set directories to be used
working_from = "charite"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/"
} else
if (working_from == "charite") {
  base_dir = "C:/MyStuff/"
}
```




# 1. Load data
```{r}
# Our proteomics data
resp <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/produced_during_analysis/responsiveness_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".csv", sep="")))

# Load their phenotypic data: growth rate and growth yield for all KOs
phenotypic_data <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Articles/bloom_ackermann_2014/phenotypic_results_2014.tsv", sep="")))
phenotypic_data_full <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Articles/bloom_ackermann_2014/phenotypic_results_full.tsv", sep="")))

# Load trna_ko
y.protein <- readRDS(paste(base_dir, "tRNA_KOs/Data/limpa_analysis/y.protein.rds", sep=""))
trna_ko <- as.data.frame(y.protein$E)
```






# 2. Correlation with the original article by Pizpel et al., 2014
## 2.1. Correlation between GR and GY measured in 2014 (published) and in 2020 (I assume they measured it before they sent us the library)
```{r}
phenotypic_data_full <- phenotypic_data_full %>%
  dplyr::select(starts_with("GR"), starts_with("GY"))

corrs <- round(cor(phenotypic_data_full, use = "pairwise.complete.obs"), 2)
melted_corrs <- melt(corrs)

ggplot(data = melted_corrs, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  geom_text(aes(Var1, Var2, label = value), size = 3) +
  scale_fill_gradient2(low = "blue", high = "red",
                       limit = c(-1,1), name="Correlation") +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```
--> Yes, correlation between those that were measured in both cases is >0.96 for all of them, so I am safe using just one of them. I will use only the ones from 2014, following Julia's recommendation, since it's published data. 


## 2.2. Correlation between our proteomics data and their phenotypic data (GR and GY)
```{r}
# Add nDEP to phenotypic data
resp_temp <- resp %>%
  dplyr::select(Strain.Name, nDEP)
test <- left_join(phenotypic_data, resp, by = "Strain.Name")

# Select numeric columns, calculate correlations and plot heatmap
test <- test %>% 
  dplyr::select(GR_YPD_2014, GR_SDC_2014, GR_low_glucose_2014, GR_galactose_2014, GR_NaCl_2014, GR_DTT_2014, GY_YPD_2014, GY_SDC_2014, GY_low_glucose_2014, GY_galactose_2014, GY_NaCl_2014, GY_DTT_2014, nDEP)
colnames(test) <- c("GR_YPD_2014", "GR_SDC_2014", "GR_low_glucose_2014", "GR_galactose_2014", "GR_NaCl_2014", "GR_DTT_2014", "GY_YPD_2014", "GY_SDC_2014", "GY_low_glucose_2014", "GY_galactose_2014", "GY_NaCl_2014", "GY_DTT_2014", "nDEP")
test_corrs <- round(cor(test, use = "pairwise.complete.obs"), 2)
melted_corrs <- melt(test_corrs)

# MY MELT FUNCTION RETURNS THE ROWS IN THE DATAFRAME IN AN ORDER THAT MAKES SENSE TO ME, BUT SEEMINLGY GGPLOT2 DOESN'T LIKE IT BECAUSE IT PLOTS THEM IN ANOTHER ORDER XD
# GOTTA FIX THAT FOR IT TO LOOK PRETTY
melted_corrs$Var1 <- factor(melted_corrs$Var1, ordered = T, levels = c("GR_YPD_2014", "GR_SDC_2014", "GR_low_glucose_2014", "GR_galactose_2014", "GR_NaCl_2014", "GR_DTT_2014", "GY_YPD_2014", "GY_SDC_2014", "GY_low_glucose_2014", "GY_galactose_2014", "GY_NaCl_2014", "GY_DTT_2014", "nDEP"))
melted_corrs$Var2 <- factor(melted_corrs$Var2, ordered = T, levels = c("GR_YPD_2014", "GR_SDC_2014", "GR_low_glucose_2014", "GR_galactose_2014", "GR_NaCl_2014", "GR_DTT_2014", "GY_YPD_2014", "GY_SDC_2014", "GY_low_glucose_2014", "GY_galactose_2014", "GY_NaCl_2014", "GY_DTT_2014", "nDEP"))

ggplot(data = melted_corrs, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  geom_text(aes(Var1, Var2, label = value), size = 3) +
  scale_fill_gradient2(low = "blue", high = "red",
                       limit = c(-1,1), name="Correlation") +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Pairwise correlations between growth data and nDEP")

# Remove unnecessary objects
#rm(test, test_corrs, melted_corrs)
```
nDEP has:
  - No correlation with growth rate under SDC, YPD and low glucose
  - Relatively large, positive correlation with growth rate and growth yield under DTT: those that show more DEPs in rich medium are those that        grow better under DTT, this agrees with the original article: these strains have proteotoxic stress going on already in the cell because of the     lack of that necessary (major?) tRNA, so they adapt better to living in DTT. 
  - Relatively large, negative correlation with growth rates and growth yields under all other conditions: also makes sense with the original            article, the strains which are less affected by the KO (so less DEPs) grow better in most conditions, simply because less things are affected     inside the cell, reasonable. The more unintuitive one is the previous one. 


## 2.3. Compare our proteomics data and their microarray data
Important to remember that the microarray data are only for 5 tRNA families: 
  - MC (multi-copy) families: tR.UCU.M2 and tH.GUG.G1, both of which had impaired or WT growth under DTT
  - SC (single-copy) families: Two viable single deletions: tR.CCU.J and tL.GAG.G, plus the initiator methionine tiM.CAU.C. They all had improved growth under DTT. 

### Loading data
```{r}
# Load their small microarray data: fold changes for all genes but only for 5 samples
microarray_data <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Articles/bloom_ackermann_2014/microarray_fold_change_data.tsv", sep="")))

# Load all S. cerevisiae genes from AllianceMine - to be able to go from standard to systematic gene names and vice versa
yeastmine <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/databases/alliancemine_results_2024-11-20T11-19-04.tsv", sep="")))

# UniProt dataset
uniprot_db <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/databases/uniprotkb_proteome_UP000002311_2025_05_24.tsv", sep="")))

# Need to change the colnames in the microarray data to use () instead of ..
microarray_data <- microarray_data %>%
  dplyr::rename(gene_names = Gene.secondaryIdentifier)
```


### Correlation heatmap between microarray and proteomics
```{r}
# Get standard protein names
microarray_data <- match_systematic_and_standard_protein_names(data = microarray_data,
                                                               uniprot = uniprot_db,
                                                               input = "systematic",
                                                               output = "standard",
                                                               simplify = FALSE,
                                                               add_extra_columns = FALSE)

# Join with our proteomics data - keep only proteins which are present in both microarrays and proteomics
trna_ko_temp <- trna_ko %>%
  dplyr::mutate(systematic_SGD_IDs = rownames(trna_ko))
proteomics_and_microarray_data <- inner_join(trna_ko_temp, microarray_data, by = "systematic_SGD_IDs") 

# Correlation heatmap between our samples 
samples_in_microarray <- colnames(microarray_data)[2:6]
proteomics_and_microarray_subset <- data.frame(matrix(nrow = nrow(proteomics_and_microarray_data), ncol = 0))

for (i in 1:length(samples_in_microarray)) {
  sample_name <- samples_in_microarray[i]
  temp <- proteomics_and_microarray_data[, grepl(sample_name, colnames(proteomics_and_microarray_data))]
  proteomics_and_microarray_subset <- cbind(proteomics_and_microarray_subset, temp)
}

cormat <- round(cor(proteomics_and_microarray_subset, use = "pairwise.complete.obs", method = "kendall"), 2)
melted_cormat <- melt(cormat)

ggplot(data = melted_cormat, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  geom_text(aes(Var2, Var1, label = value), size = 3) +
  scale_fill_gradient2(low = "blue", high = "red",
                       limit = c(-1,1), name="Correlation") +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Pairwise correlations between transcriptomics and proteomics")
```
--> The proteomics samples for these 5 strains have literally 0 correlation with the microarray data from the same strains. Could it be because the proteomics data are direct measurements while the microarray data are actually presented as the fold-change with respect to the WT? I don't think so but maybe I could look into it? Could try to make my proteomics data for these strains into that as well?


### Venn diagram between genes detected in proteomics and in transcriptomics
```{r}
list_for_venn_diagram <- list(Proteomics = rownames(trna_ko),
                              Transcriptomics = microarray_data$systematic_SGD_IDs)
ggvenn(
  list_for_venn_diagram, 
  fill_color = c("#80AAFF", "#FF8080"),
  stroke_size = 0.5, set_name_size = 4
  )
```


### Correlation heatmap within microarray
```{r}
test <- microarray_data %>% 
  dplyr::select(-systematic_SGD_IDs, -standard_IDs)
corrs <- round(cor(test, use = "pairwise.complete.obs"), 2)
melted_corrs <- melt(corrs)

ggplot(data = melted_corrs, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  geom_text(aes(Var2, Var1, label = value), size = 3) +
  scale_fill_gradient2(low = "blue", high = "red",
                       limit = c(-1,1), name="Correlation") +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Pairwise correlations between transcriptomics fold changes w.r.t. WT")

# Remove unnecessary variables
rm(test, corrs, melted_corrs)
```


### Correlation heatmap within proteomics (only the samples involved here)
```{r}
# Grab columns of proteomics data with only samples from the strains for which there are microarray data
matches <- grepl(paste(samples_in_microarray, collapse="|"), 
                        colnames(trna_ko))
MC_and_SC_proteomics <- trna_ko[, matches]

# Add MC or SC information to column names and sort columns by this so (expected) clusters are easily visible on heatmap
new_colnames <- c()
for (i in 1:ncol(MC_and_SC_proteomics)) {
  old_colname <- colnames(MC_and_SC_proteomics)[i]
  if (grepl(paste(c("UCU", "GUG"), collapse = "|"), old_colname))
    new_colnames <- c(new_colnames, paste("MC", old_colname, sep = "_"))
  else if ((grepl(paste(c("CCU", "GAG", "CAU"), collapse = "|"), old_colname)))
    new_colnames <- c(new_colnames, paste("SC", old_colname, sep = "_"))
}
colnames(MC_and_SC_proteomics) <- new_colnames
MC_and_SC_proteomics <- MC_and_SC_proteomics %>% dplyr::select(order(colnames(MC_and_SC_proteomics)))

# Correlation heatmap
corrs <- round(cor(MC_and_SC_proteomics, use = "pairwise.complete.obs"), 2)
melted_corrs <- melt(corrs)

ggplot(data = melted_corrs, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  geom_text(aes(Var2, Var1, label = value), size = 3) +
  scale_fill_gradient2(low = "blue", high = "red",
                       limit = c(-1,1), name="Correlation") +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))



# Remove unnecessary objects
rm(toMatch, matches, corrs, melted_corrs, old_colname, new_colnames)
```
--> In the heatmap based on microarray data, the SC and MC samples separate quite nicely, each of them correlating only with those in the same group as itself. However, at the proteomic level we see extremely high correlation between all of the samples, with no particular pattern to be seen. Is this something to do with how the proteomics data were processed? Or?
Okay this probably also explains why I see no correlation at all between the proteomics data and the microarray in the other heatmap, but still, what is going on here???

Get one of these for all the proteomics data and see if they all are this heavily correlated to each other or if I can actually see some pattern per KOs. Because if all of them are this correlated to each other, it feels a bit wrong I'd say? Idk, maybe it's normal. Also maybe try to get a heatmap for the proteomics data themselves, not only for the correlations between them


### Correlation heatmap of only proteomics data (for all samples) - pretty useless
```{r}
corrs <- round(cor(trna_ko, use = "pairwise.complete.obs"), 2)
melted_corrs <- melt(corrs)

ggplot(data = melted_corrs, aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile() +
  scale_fill_gradient2(low = "blue", high = "red",
                       limit = c(-1,1), name="Correlation") +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        axis.text.x=element_blank(), 
        axis.ticks.x=element_blank(), 
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())
```



# 3. Check that proteins which are upregulated in these samples in the proteomics data, are also upregulated in microarray data - if this is the case, it should mean that my correlation between the 2 datasets is not wrong and that in fact there are interesting things to look at here
```{r}
# Load da dataframe
da <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/da_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".csv", sep="")))

microarray_samples <- colnames(microarray_data[2:6])

da_for_microarray_samples <- da %>% 
  dplyr::filter(Strain.Name %in% microarray_samples)

for (i in 1:length(microarray_samples)) {
  sample <- microarray_samples[i]
  da_temp <- da %>% 
    dplyr::filter(Strain.Name == sample) %>%
    dplyr::select(protein, diffexpressed_adjusted, logFC, adj.P.Val)
  colnames(da_temp) <- c("systematic_SGD_IDs", "diffexpressed", "logFC", "adj.P.Val")
  microarray_temp <- microarray_data %>% 
    dplyr::select(all_of(sample), systematic_SGD_IDs)
  test <- left_join(microarray_temp, da_temp, by = "systematic_SGD_IDs")
  test <- inner_join(da_temp, microarray_temp, by = "systematic_SGD_IDs")
  colnames(test) <- c("systematic_SGD_IDs", "diffexpressed", "logFC", "adj.P.Val", "logfold_microarray")
  
  
  # Calculate correlation to add to plot
  corr_coef <- round(cor(test$logFC, test$logfold_microarray, method = "spearman", use = "complete.obs"), 2)
  
  
  # Protein names in X axis, transcriptomics fold change in Y axis, color by significance in proteomics - not such a good plot tbh
  
  #p1 <- ggplot(data = test, aes(x = Gene.symbol, y = as.numeric(logfold_microarray), col = diffexpressed)) +
  #  geom_point() +
  #  labs(title = sample,
  #       col = "Differentially\nexpressed\nin proteomics\ndata") +
  #  xlab("Protein name") +
  #  ylab("Transcriptomics fold change") +
  #  theme(axis.text.x=element_blank(),
  #      axis.ticks.x=element_blank()) +
  #  geom_text(data = subset(test, diffexpressed == "Up_regulated" | diffexpressed == "Down_regulated"),
  #            aes(x = Gene.symbol, y = as.numeric(logfold_microarray), col = diffexpressed, label = Gene.symbol),
  #            size = 5)
  #print(p1)
  
  
  # logFC transcriptomics vs. logFC proteomics - color by significance in proteomics: much better
  p2 <- ggplot(data = test, aes(x = logFC, y = logfold_microarray, col = diffexpressed)) +
    geom_point(size = 1.5) +
    labs(title = glue('{sample} - correlation coef. = {corr_coef}'),
         col = "Differentially\nexpressed\nin proteomics\ndata") +
    xlab("logFC proteomics") +
    ylab("logFC transcriptomics") +
    theme_light() +
    geom_text_repel(data = subset(test, diffexpressed == "Up_regulated" | diffexpressed == "Down_regulated"),
              aes(x = logFC, y = logfold_microarray, col = diffexpressed, label = systematic_SGD_IDs),
              size = 4) +
    geom_abline() +
    scale_color_manual(values = c(Down_regulated = "#80AAFF", Not_significant = "#808080", Up_regulated = "#FF8080"))
  print(p2)
  
  # This is just to check if the FCs from the microarrays data are also log2FCs, which I think so but their article says only FC
  #hist(test$logfold_microarray)
  #hist(test$logFC)
  #hist((test$logFC)^2)
}

```
--> The correlations here get quite a bit better than in the heatmap for proteomics vs. transcriptomics, I guess because the proteomics data is averaged over the replicates for each tRNA? That's the only difference I can think of

--> Interestingly, I get decent correlations between proteomics and transcriptomics in the KOs for MC tRNAs, those which grew worse than WT or equally bad in DTT, so those where we assume the removal of the tRNA is not having a proteotoxic effect. On the other hand, for SC tRNAs (those where the removal of the tRNA has a proteotoxic effect, hence allowing them to grow better than WT in DTT), there is barely any correlation between transcriptomics and proteomics. 






























