---
title: "enrichment_analysis"
author: "Álvaro Gómez Pérez"
date: "2025-02-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries
```{r}
library(clusterProfiler)
library(org.Sc.sgd.db)
library(data.table)
library(dplyr)
library(fgsea)
library(ggplot2)
library(jsonlite)
library(stringr)
library(forcats)
library(GO.db)
library(biomaRt)
```



# 1. Load and prepare data
## 1.0. Set parameters
```{r}
# Significance level used for defining proteins as DE
alpha <- 0.05

# Significance level above as a plain string to use when loading or writing files
alpha_plain <- str_replace(as.character(alpha), "\\.", "")

# Separate significance level, the one used for the enrichment analysis
alpha_enrichment <- 0.1

# Significance level above as a plain string to use when loading or writing files
alpha_enrichment_plain <- str_replace(as.character(alpha_enrichment), "\\.", "")

# Log-fold change limit to be considered "biologically significant"
lfc_threshold <- 0.5

# Number of nDEP from which a tRNA is considered "major" (this threshold value should be included) - THIS I AM NOT USING ANYMORE, AM I?
threshold_major <- 2

# Set directories to be used
working_from = "home"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/"
} else
if (working_from == "charite") {
  base_dir = "C:/MyStuff/"
}
```

## 1.1. Load data
```{r}
# List of vectors with the proteins that were DE in each KO strain - remove emtpy vectors
de_proteins_list <- fromJSON(paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/de_proteins_list_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep=""), )
de_proteins_list <- Filter(function(x) length(x) > 0, de_proteins_list)

# Full Entrez ID dataframe, with all genes
entrez_db <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/entrez_reference.txt", sep="")))

# Load the dataset with that information
master_dataset <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/basic/master_tRNA_dataset.csv", sep="")))

# Phenotypic data from Bloom-Ackermann et al., 2014
phenotypic_data <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Articles/bloom_ackermann_2014/phenotypic_results_full.tsv", sep="")))


source(paste(base_dir, "tRNA_KOs/Code/R/Mine/0.general_use_functions.R", sep=""))
```



# 2. GO enrichment analysis - for proteins at 0.01 - THE DATA PRODUCED FOR THIS ARE NOT UPDATED, THEY ARE NOT PRODUCED BY THE NEW FUNCTIONS I USE FOR DA! 
I COULD ADAPT THEM FOR THAT, BUT I THINK I CAN USE GSEA WITH GO ANNOTATION INSTEAD??? IN THAT WAY I CAN JUST WORK FROM THE DA DATAFRAME

## 2.1. Get NCBI gene IDs for each of the systematic name vectors I have
```{r}
warning("THIS CODE IS NOT SUPPOSED TO RAN ANYMORE! READ THE TITLE OF THE SECTION")
de_proteins_list_entrez_ID <- list()
for (i in 1:length(de_proteins_list)) {
  temp <- data.frame("Protein_stable_ID" = de_proteins_list[[i]])
  temp <- left_join(temp, entrez_db, by = "Protein_stable_ID")
  temp <- temp$`NCBI_gene_(formerly_Entrezgene)_ID`
  de_proteins_list_entrez_ID[[i]] <- temp
}
names(de_proteins_list_entrez_ID) <- names(de_proteins_list)
```

## 2.2. Get the information of how many DE proteins there are for each tRNA KO strain
```{r}
de_prots_per_strain <- c()
for (i in 1:length(de_proteins_list)) {
  temp <- de_proteins_list[[i]]
  de_prots_per_strain <- c(de_prots_per_strain, length(temp))
}

de_prots_per_strain <- data.frame(Strain.Name = names(de_proteins_list),
                                      nDEP = de_prots_per_strain) %>%
  filter(Strain.Name != "WT")
```

## 2.3. Perform GO enrichment analysis
```{r}
# Come up with background (all proteins detected)
my_universe <- as.character(de_proteins_list_entrez_ID[["WT"]])

# Create empty list for GO results, perform GO for each KO strain, and store them there
go_results_list <- list()
for (i in 2:length(de_proteins_list_entrez_ID)) {
  proteins_detected <- as.character(de_proteins_list_entrez_ID[[i]])
  if (length(proteins_detected) > 0) {
    go_results <- enrichGO(gene = proteins_detected, 
                           OrgDb = "org.Sc.sgd.db", 
                           keyType = "ENTREZID", 
                           ont = "ALL", 
                           universe = my_universe)
    go_results <- as.data.frame(go_results)
    if (nrow(go_results) > 0) {
      go_results_list[[names(de_proteins_list_entrez_ID)[i]]] <- go_results
    }
  }
}

# Check how much correlation there is between strains having many DEPs and strains showing some GO enrichment results
has_go_results <- c()
number_of_go_terms <- c()
for (i in 1:nrow(de_prots_per_strain)) {
  strain <- de_prots_per_strain$Strain.Name[i]
  if (strain %in% names(go_results_list)) {
    has_go_results <- c(has_go_results, "Yes")
    number_of_go_terms <- c(number_of_go_terms, nrow(go_results_list[[strain]]))
  }
  else {
    has_go_results <- c(has_go_results, "No")
    number_of_go_terms <- c(number_of_go_terms, 0)
  }
}
de_prots_per_strain$GO_results_bp <- has_go_results
de_prots_per_strain$num_of_go_terms_bp <- number_of_go_terms
```

## 2.4. Add other informational columns to this dataset that might explain the success of the GO analysis for some strains
```{r}
de_prots_per_strain <- left_join(de_prots_per_strain, master_dataset, by = "Strain.Name")
```

## 2.5. Save the EA dataset so I don't have to run all of this again
```{r}
write_json(go_results_list, path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_results_list_", alpha_plain, "_logFC_", as.character(lfc_threshold),  ".json", sep=""), pretty = T)
```







# 3. Evaluate overall results of EA before going into strain comparisons
## 3.1. Load data
```{r}
go_results_list <- read_json(path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_results_list_", alpha_plain, "_logFC_", as.character(lfc_threshold),  ".json", sep=""), simplifyVector = T)
```

## 3.2. Barplots for most common EA terms
```{r}
# Prepare dataframe
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_results_list <- mapply(FUN = add_strain_tag, df = go_results_list, strain = names(go_results_list), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_full <- bind_rows(go_results_list) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))


# Plot
## Grab only top 10 most common EA terms
temp <- go_terms_df_full %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```









# 4. Compare EA results to growth in DTT (equal or worse than WT vs. better than WT) 
## 4.1. Load data
```{r}
# Results from GO EA
go_results_list <- read_json(path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_results_list_", alpha_plain, "_logFC_", as.character(lfc_threshold),  ".json", sep=""), simplifyVector = T)

# Load the dataset with that information
master_dataset <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/basic/master_tRNA_dataset.csv", sep="")))

# Phenotypic data from Bloom-Ackermann et al., 2014
phenotypic_data <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Articles/bloom_ackermann_2014/phenotypic_results_full.tsv", sep="")))

source(paste(base_dir, "tRNA_KOs/Code/R/Mine/0.general_use_functions.R", sep=""))
```

## 4.2. Evaluate growth in DTT and classify strains to better than WT or equal/worse
```{r}
# Plot GR and GY in DTT
ggplot(data = phenotypic_data, aes(x = DTT)) +
  geom_histogram(binwidth = 1) +
  theme_light() +
  geom_vline(xintercept = mean(phenotypic_data$DTT, na.rm = T)+sd(phenotypic_data$DTT, na.rm = T), col = "red") +
  geom_vline(xintercept = mean(phenotypic_data$DTT, na.rm = T)-sd(phenotypic_data$DTT, na.rm = T), col = "red")

ggplot(data = phenotypic_data, aes(x = DTT_GY)) +
  geom_histogram(binwidth = 1) +
  theme_light() +
  geom_vline(xintercept = mean(phenotypic_data$DTT_GY, na.rm = T)+sd(phenotypic_data$DTT_GY, na.rm = T), col = "red") +
  geom_vline(xintercept = mean(phenotypic_data$DTT_GY, na.rm = T)-sd(phenotypic_data$DTT_GY, na.rm = T), col = "red")

# At first I just divided strains into those above and below 0, and grabbed those that were above 0 in both GR and GY as one group, and those that were below 0 in both GR and GY as the other group. However, I found no pattern at all, so I think it would make sense to be a bit more stringent in how I select the strains. I am going to do it based on the SD of each of the distribution above

# Need to look into the original article to figure out which one to use - going with both for now, so I don't take all strains here, only those that agree on their GR and GY (not sure if that makes sense, going by vibes for now)
## Need to get rid of the lethal strains before I select here
phenotypic_data <- phenotypic_data %>%
  filter(lethal != "Yes",
         !(is.na(DTT)),
         !is.na(DTT_GY))

## Actually perform the selection
strains_better_in_DTT <- phenotypic_data$gene_name[phenotypic_data$DTT > sd(phenotypic_data$DTT, na.rm = T) & phenotypic_data$DTT_GY > sd(phenotypic_data$DTT_GY, na.rm = T)]
strains_worse_or_equal_in_DTT <- phenotypic_data$gene_name[phenotypic_data$DTT < -sd(phenotypic_data$DTT, na.rm = T) & phenotypic_data$DTT_GY < -sd(phenotypic_data$DTT_GY, na.rm = T)]
```

## 4.3. Compare EA results to these strain classification
```{r}
go_list_better_dtt <- list()
go_list_worse_dtt <- list()

for (i in 1:length(go_results_list)) {
  strain_name <- names(go_results_list)[i]
  if (strain_name %in% strains_better_in_DTT) {
    go_list_better_dtt[[strain_name]] <- go_results_list[[i]]
  }
  else if (strain_name %in% strains_worse_or_equal_in_DTT) {
    go_list_worse_dtt[[strain_name]] <- go_results_list[[i]]
  }
}
```

## 4.4. Come up with a count of the enriched pathways throughout each set of strains
Get dataframes
```{r}
# For strains growing worse in DTT
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_list_worse_dtt <- mapply(FUN = add_strain_tag, df = go_list_worse_dtt, strain = names(go_list_worse_dtt), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_worse_dtt <- bind_rows(go_list_worse_dtt) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
  

# For strains growing better in DTT
## Add strain tag to each dataframe before collapsing
go_list_better_dtt <- mapply(FUN = add_strain_tag, df = go_list_better_dtt, strain = names(go_list_better_dtt), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_better_dtt <- bind_rows(go_list_better_dtt) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
```

Plot
```{r}
# For strains growing worse in DTT
## Grab only top 10 most common EA terms
temp <- go_terms_df_worse_dtt %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")




# For strains growing better in DTT
## Grab only top 10 most common EA terms
temp <- go_terms_df_better_dtt %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```

--> Pretty much the same in both subsets, let's see what is actually different between them
```{r}
# Filter the dataframes
ea_only_worse_df <- go_terms_df_worse_dtt[!(go_terms_df_worse_dtt$Description %in% go_terms_df_better_dtt$Description),]
ea_only_better <- go_terms_df_better_dtt[!(go_terms_df_better_dtt$Description %in% go_terms_df_worse_dtt$Description),]



# Plot
# For strains growing worse in DTT
## Grab only top 10 most common EA terms
temp <- ea_only_worse_df %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")




# For strains growing better in DTT
## Grab only top 10 most common EA terms
temp <- ea_only_better %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing better than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```










# 5. Gene Set Enrichment Analysis (GSEA)
In order to perform GSEA with the gsea() function from the gsea package in R, we need the following:
  - A list of named vectors, where the names are those of the gene sets we are working with (GO, KEGG, whatever) and the corresponding vector contains the names of the genes that are included in that gene set. Only genes that are detected in our dataset should be included here (this is at the same time the reference and the background for our enrichment analysis).
  - A named vector of ranked genes - we calculate ranks/scores for the genes detected in each strain, which represent how DE they are basically. The values in this vector are these ranks/scores, while their names are simply the name of the gene. 
  
That's it! We provide gsea() with these things and with some other parameters, and it will perform the enrichment analysis :)



## 5.1. Prepare ranked gene lists
For now, my score/rank is: sign(logFC)*(-log10(adj.P.Val))
```{r}
# Load list of all dataframes (one per strain) with the DE results
da <- read_json(paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/da_", alpha_plain, "_logFC_", as.character(lfc_threshold), ".json", sep=""), simplifyVector = T)

# Now da is a collapsed dataframe with all strains, need to separate it per strain and put the resulting dataframes into a list
da_list <- list()
strains <- unique(da$Strain.Name)
for (i in 1:length(strains)) {
  strain <- strains[i]
  temp <- da %>%
    dplyr::filter(Strain.Name == strain)
  da_list[[strain]] <- temp
}
da <- da_list

# Create empty list to put the named ranked gene lists in
ranked_genes <- list()
number_of_proteins_with_no_Entrez_ID <- c()

# Iterate over these dataframes
for (i in 1:length(da)) {
  temp <- da[[i]]
  
  # Get the corresponding Entrez IDs to these systematic gene names, since the GSEA is performed in terms of those
  temp$Protein_stable_ID <- rownames(temp)
  entrez_temp <- entrez_db %>%
    dplyr::select(Protein_stable_ID, `NCBI_gene_(formerly_Entrezgene)_ID`)
  temp <- left_join(temp, entrez_temp, by = "Protein_stable_ID")
  rownames(temp) <- NULL
  temp <- temp %>%
    dplyr::select(logFC, adj.P.Val, `NCBI_gene_(formerly_Entrezgene)_ID`) %>%
    dplyr::rename(gene_symbol = `NCBI_gene_(formerly_Entrezgene)_ID`)
  
  # Get rid of the proteins for which we don't have the Entrez ID (and save the count of how many of them there are, to check it's not too many) - otherwise GSEA doesn't work
  number_of_proteins_with_no_Entrez_ID <- c(number_of_proteins_with_no_Entrez_ID, sum(is.na(temp$gene_symbol)))
  temp <- temp %>%
    dplyr::filter(!is.na(gene_symbol))
  
  # Using as score the -log10(p-value)*sign of the logFC
  rankings <- sign(temp$logFC)*(-log10(temp$adj.P.Val))
  names(rankings) <- temp$gene_symbol
  rankings <- sort(rankings, decreasing = TRUE)
  
  # Plot
  plot(rankings)
  
  # Check max and min - to see if there is a need to correct for -Inf and +Inf - commented out, I don't think it's necessary
  #print(max(rankings))
  #print(min(rankings))
  
  # Some genes have such low p values that the signed pval is +- inf, we need to change it to the maximum * constant to avoid problems with fgsea
  # (this part is copy pasted from the link)
  max_ranking <- max(rankings[is.finite(rankings)])
  min_ranking <- min(rankings[is.finite(rankings)])
  rankings <- replace(rankings, rankings > max_ranking, max_ranking * 10)
  rankings <- replace(rankings, rankings < min_ranking, min_ranking * 10)
  rankings <- sort(rankings, decreasing = TRUE) # sort genes by ranking
  
  # Add them to the list where we are saving them
  ranked_genes[[names(da)[i]]] <- rankings
}


# Save this ranked lists so that I can use them later to pick the top 50 proteins in each strain, to look at the overlaps between major strains within families in a more unbiased manner - not really feasible for now since the distribution of the scores for the vast majority of strains is pretty much flat. I wonder how much this affects my DEA?
write_json(ranked_genes, path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/lists_of_ranked_genes_", alpha_plain, "_logFC_", as.character(lfc_threshold), sep=""),
           pretty = T)
```



## 5.2. Perform GSEA with KEGG annotation - as described by https://biostatsquid.com/fgsea-tutorial-gsea/
### Download KEGG gene lists (according to https://www.researchgate.net/post/How_i_can_get_a_list_of_KEGG_pathways_and_its_list_of_genes)
DO NOT RUN AGAIN - I used this to download the .gmt file, I just need to load it, but I left this code here in case I want to know how I did it
```{r}
#BiocManager::install("KEGGREST")
#BiocManager::install("EnrichmentBrowser")
#library("KEGGREST")
#library("EnrichmentBrowser")
#
##step2: check and obtain a list of entry identifiers (in this case: sce) and associated definition for a given database or a given set of database entries.
#MRSA252 <- keggList("sce")
#
##step 3: download the pathways of that organism:
#sarpathway <- downloadPathways("sce")
#
##step 4: retrieve gene sets for an organism from databases such as GO and KEGG:
#sar <- getGenesets(org = "sce", db = "kegg", cache = TRUE, return.type="list")
#
##step5: Parse and write the gene sets to a flat text file in GMT format for other pathway enrichment analysis programs (e.g., GSEA):
#writeGMT(sar, gmt.file = paste(base_dir, "Data/Other/enrichment_analysis/20250305_kegg_sce_gmt", sep=""))
```

### Process .gmt file to turn it into the list of gene set vectors we need for gsea()
Get a list of pathways, with each entry being a vector with the Entrez IDs for the genes involved in that pathway. Make it so that only proteins that we detect in our dataset are included here. 
```{r}
# Read in the .gmt file
gmt <- read.gmt(paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/20250305_kegg_sce_gmt", sep=""))
gmt$term <- as.character(gmt$term)

# Get a simple vector with the Entrez IDs of the proteins we detect
trna_ko <- as.data.frame(as.matrix(fread(paste(base_dir, "tRNA_KOs/Data/proteomics_data/processed_proteomics_dataframe.tsv", sep="")), rownames = 1))
yeastmine <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/databases/alliancemine_results_2024-11-20T11-19-04.tsv", sep="")))
proteins_detected <- match_systematic_and_standard_protein_names(data = rownames(trna_ko),
                                                                 yeastmine = yeastmine,
                                                                 input = "standard",
                                                                 simplify = T)
proteins_detected <- entrez_db$`NCBI_gene_(formerly_Entrezgene)_ID`[entrez_db$Protein_stable_ID %in% proteins_detected]

# Get our list of vectors, where each vector is a KEGG pathway and the terms in it are the genes related to it - keeping only genes which are detected in our dataset
gmt <- gmt %>%
  mutate(pathway_code = substr(term, 1, str_locate(term, "_")[1]-1), 
         pathway_name = substr(term, str_locate(term, "_")[1]+1, nchar(term)))

kegg_genes <- list()
pathways <- unique(gmt$pathway_name)
for (i in 1:length(pathways)) {
  pathway <- pathways[i]
  temp <- gmt %>%
    filter(pathway_name == pathway)
  kegg_genes[[pathway]] <- as.character(temp$gene)[as.character(temp$gene) %in% proteins_detected]
}
```

### Run GSEA
The warning I get here refers to many proteins having exactly the same rank (these are the ones we see in a horizontal line in the plots produced in 5.1.) - this should be okay, since these are the genes that were not significantly differentially expressed at all - but tbh this makes it so that I cannot really get that much info from those anyway, right? So there isn't that much of a difference between this and doing the GSEA only with DE proteins, right?? Idk
```{r}
kegg_gsea_results <- list()

for (i in 1:length(ranked_genes)) {
  kegg_gsea_results[[names(ranked_genes)[i]]] <- fgsea(pathways = kegg_genes,                                # List of gene sets to check
                                                       stats = ranked_genes[[i]],
                                                       scoreType = 'std',                                     
                                                                      # in this case we have both pos and neg rankings. if only pos or neg, set to 'pos', 'neg'
                                                       minSize = 1,
                                                       maxSize = 1000,
                                                       nproc = 1)                                             # for parallelisation
}


# This returns a list, where each element is a dataframe for the corresponding KO strain, and each row in that dataframe is a pathway. However, all pathways are included, so we are going to filter all of these dataframes to keep only the pathways that are significant in each KO strain
kegg_gsea_results_filtered <- lapply(kegg_gsea_results, function(x) filter(x, padj < alpha_enrichment))
bool <- sapply(kegg_gsea_results_filtered, function(x) nrow(x)>0)
kegg_gsea_results_filtered <- kegg_gsea_results_filtered[bool]

# Save these results
write_json(kegg_gsea_results_filtered, paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/kegg_gsea_results_", alpha_plain, "_logFC_", as.character(lfc_threshold), "_enrichment_alpha_", alpha_enrichment_plain, ".json", sep=""), pretty = T)
```



## 5.3. GSEA with GO annotation
### Get the information equivalent to that in the .gmt file and save it as gene set vectors already - code from ChatGPT
Initially I was then saving this info to an actual .gmt file to them use it, but honestly that just made everything more convoluted, I'm just going to save it as a list of vectors, as a JSON file and that's it :)
```{r}
# Step 1: Connect to Ensembl Fungi
ensembl_fungi <- useMart("fungi_mart", dataset = "scerevisiae_eg_gene", host = "https://fungi.ensembl.org")

# Step 2: Retrieve GO annotations
go_annotations <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", 
                                       "go_id", "name_1006", "namespace_1003", "entrezgene_id"),
                        mart = ensembl_fungi)

# This is added by me
trna_ko <- as.data.frame(as.matrix(fread(paste(base_dir, "tRNA_KOs/Data/proteomics_data/processed_proteomics_dataframe.tsv", sep="")), rownames = 1))
go_annotations <- go_annotations %>%
  dplyr::filter(namespace_1003 != "go") %>%                                     # Pretty sure these rows are just wrong, or at least not useful to me
  dplyr::filter(ensembl_gene_id %in% rownames(trna_ko)) %>%                     # Keep only those genes which are detected in our data, so as to prevent bias
  dplyr::mutate(gene_set = case_when(namespace_1003 == "molecular_function" ~ paste(name_1006, " (MF)", sep=""),
                              namespace_1003 == "biological_process" ~ paste(name_1006, " (BP)", sep=""),
                              namespace_1003 == "cellular_component" ~ paste(name_1006, " (CC)", sep="")))

# Check the first few rows
head(go_annotations)

# Step 3: Filter for Biological Process (BP) only - not interested in using this for now
go_annotations <- go_annotations %>%
  #filter(namespace_1003 == "biological_process") %>%
  filter(go_id != "")  # Remove any empty GO terms if present

# Step 4: Build a list: GO Term -> Genes
# Use external_gene_name (common names) or ensembl_gene_id (your choice)
go_list <- split(go_annotations$entrezgene_id, go_annotations$gene_set)

# Step 5: Save this named list of vectors as a JSON file
write_json(go_list, path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_list_gmt.json", sep=""), pretty = T)
```

### Load the gene set vectors created in the previous chunk
Get a list of pathways, with each entry being a vector with the Entrez IDs for the genes involved in that pathway. Make it so that only proteins that we detect in our dataset are included here. 
```{r}
# Read in the .json file
go_gene_sets <- read_json(path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_list_gmt.json", sep=""), simplifyVector = T)
```

### Run GSEA
The warning I get here refers to many proteins having exactly the same rank (these are the ones we see in a horizontal line in the plots produced in 5.1.) - this should be okay, since these are the genes that were not significantly differentially expressed at all - but tbh this makes it so that I cannot really get that much info from those anyway, right? So there isn't that much of a difference between this and doing the GSEA only with DE proteins, right?? Idk
```{r}
go_gsea_results <- list()

for (i in 1:length(ranked_genes)) {
  go_gsea_results[[names(ranked_genes)[i]]] <- fgsea(pathways = go_gene_sets,                                # List of gene sets to check
                                                       stats = ranked_genes[[i]],
                                                       scoreType = 'std',                                     
                                                                      # in this case we have both pos and neg rankings. if only pos or neg, set to 'pos', 'neg'
                                                       minSize = 1,
                                                       maxSize = 1000,
                                                       nproc = 1)                                             # for parallelisation
}


# This returns a list, where each element is a dataframe for the corresponding KO strain, and each row in that dataframe is a pathway. However, all pathways are included, so we are going to filter all of these dataframes to keep only the pathways that are significant in each KO strain
go_gsea_results_filtered <- lapply(go_gsea_results, function(x) filter(x, padj < alpha_enrichment))
bool <- sapply(go_gsea_results_filtered, function(x) nrow(x)>0)
go_gsea_results_filtered <- go_gsea_results_filtered[bool]

# Save these results
write_json(go_gsea_results_filtered, paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_gsea_results_", alpha_plain, "_logFC_", as.character(lfc_threshold), "_enrichment_alpha_", alpha_enrichment_plain, ".json", sep=""), pretty = T)
```




Perform GSEA with Reactome annotation?? - Seeminlgy Reactome is mostly human, so if anything I could look to repeat this with GO annotation, but not Reactome!



## 5.4. Analyze GSEA results
### 5.4.1. Load results
```{r}
kegg_gsea_results <- read_json(path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/kegg_gsea_results_", alpha_plain, "_logFC_", as.character(lfc_threshold), "_enrichment_alpha_", alpha_enrichment_plain, ".json", sep=""), simplifyVector = T)
go_gsea_results <- read_json(path = paste(base_dir, "tRNA_KOs/Data/enrichment_analysis/go_gsea_results_", alpha_plain, "_logFC_", as.character(lfc_threshold), "_enrichment_alpha_", alpha_enrichment_plain, ".json", sep=""), simplifyVector = T)
```


### 5.4.2. Barplots for most common GSEA terms
KEGG
```{r}
# Prepare dataframe
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
kegg_gsea_results <- mapply(FUN = add_strain_tag, df = kegg_gsea_results, strain = names(kegg_gsea_results), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
kegg_terms_df_full <- bind_rows(kegg_gsea_results) %>%
  summarize_mine(column_name = "pathway", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))


# Plot
## Grab only top 10 most common EA terms
temp <- kegg_terms_df_full %>%
  dplyr::distinct(pathway, .keep_all = T) %>%
  dplyr::slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_count)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common KEGG GSEA terms across all strains",
       subtitle = glue("alpha = {alpha}, logFC = {lfc_threshold}, alpha for GSEA = {alpha_enrichment}")) +
  ylab("Strain count") +
  xlab("GSEA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_percentage)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common KEGG GSEA terms across all strains",
       subtitle = glue("alpha = {alpha}, logFC = {lfc_threshold}, alpha for GSEA = {alpha_enrichment}")) +
  ylab("Strain percentage") +
  xlab("GSEA term")
```

GO
```{r}
# Prepare dataframe
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_gsea_results <- mapply(FUN = add_strain_tag, df = go_gsea_results, strain = names(go_gsea_results), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_full <- bind_rows(go_gsea_results) %>%
  summarize_mine(column_name = "pathway", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))


# Plot
## Grab only top 10 most common EA terms
temp <- go_terms_df_full %>%
  distinct(pathway, .keep_all = T) %>%
  dplyr::slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_count)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common GO GSEA terms",
       subtitle = "Across all strains") +
  ylab("Strain count") +
  xlab("GSEA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_percentage)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common GO GSEA terms",
       subtitle = "Across all strains") +
  ylab("Strain percentage") +
  xlab("GSEA term")
```


### 5.4.3. Barplot separating strains growing better or worse than WT in DTT
Load data
```{r}
# Load the dataset with that information
master_dataset <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/basic/master_tRNA_dataset.csv", sep="")))

# Phenotypic data from Bloom-Ackermann et al., 2014
phenotypic_data <- as.data.frame(fread(paste(base_dir, "tRNA_KOs/Data/Articles/bloom_ackermann_2014/phenotypic_results_full.tsv", sep="")))
```

Classify strains
```{r}
# Plot GR and GY in DTT
ggplot(data = phenotypic_data, aes(x = GR_DTT_2014)) +
  geom_histogram(binwidth = 1, fill = "grey", col = "black") +
  theme_light() +
  geom_vline(xintercept = mean(phenotypic_data$GR_DTT_2014, na.rm = T)+sd(phenotypic_data$GR_DTT_2014, na.rm = T), col = "red", linewidth = 1) +
  geom_vline(xintercept = mean(phenotypic_data$GR_DTT_2014, na.rm = T)-sd(phenotypic_data$GR_DTT_2014, na.rm = T), col = "red", linewidth = 1) +
  geom_vline(xintercept = 2, col = "darkgreen", linewidth = 1) +
  geom_vline(xintercept = -2, col = "darkgreen", linewidth = 1)

ggplot(data = phenotypic_data, aes(x = GY_DTT_2014)) +
  geom_histogram(binwidth = 1, fill = "grey", col = "black") +
  theme_light() +
  geom_vline(xintercept = mean(phenotypic_data$GY_DTT_2014, na.rm = T)+sd(phenotypic_data$GY_DTT_2014, na.rm = T), col = "red", linewidth = 1) +
  geom_vline(xintercept = mean(phenotypic_data$GY_DTT_2014, na.rm = T)-sd(phenotypic_data$GY_DTT_2014, na.rm = T), col = "red", linewidth = 1) +
  geom_vline(xintercept = 2, col = "darkgreen", linewidth = 1) +
  geom_vline(xintercept = -2, col = "darkgreen", linewidth = 1)


# Selection of strains into "better", "worse" or "un-affected" w.r.t. to WT in DTT is now done in the script where I create the master_dataframe, and it's done on a threshold of 2 and -2, which is the one they effectively use in the article (because I haven't managed to understand yet how they got these data)

## Actually perform the selection, based on the master_dataset - I take as "better" any that is actually better in either GR or GY, but not worse in either of them - same for "worse"
strains_better_in_DTT <- master_dataset$Strain.Name[(master_dataset$GR_in_DTT_compared_to_WT == "Better" & master_dataset$GY_in_DTT_compared_to_WT != "Worse") |
                                                      (master_dataset$GR_in_DTT_compared_to_WT != "Worse" & master_dataset$GY_in_DTT_compared_to_WT == "Better")]

strains_worse_in_DTT <- master_dataset$Strain.Name[(master_dataset$GR_in_DTT_compared_to_WT == "Worse" & master_dataset$GY_in_DTT_compared_to_WT != "Better") |
                                                      (master_dataset$GR_in_DTT_compared_to_WT != "Better" & master_dataset$GY_in_DTT_compared_to_WT == "Worse")]
```

KEGG - separate GSEA results based on this classification
```{r}
# Step 1 - Separate results
kegg_list_better_dtt <- list()
kegg_list_worse_dtt <- list()

for (i in 1:length(kegg_gsea_results)) {
  strain_name <- names(kegg_gsea_results)[i]
  if (strain_name %in% strains_better_in_DTT) {
    kegg_list_better_dtt[[strain_name]] <- kegg_gsea_results[[i]]
  }
  else if (strain_name %in% strains_worse_in_DTT) {
    kegg_list_worse_dtt[[strain_name]] <- kegg_gsea_results[[i]]
  }
}



# Step 2 - Turn to dataframes
## For strains growing worse in DTT
### Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
kegg_list_worse_dtt <- mapply(FUN = add_strain_tag, df = kegg_list_worse_dtt, strain = names(kegg_list_worse_dtt), SIMPLIFY = F)

### Collapse list of dataframes into single dataframe and add a column with the count of pathway
kegg_terms_df_worse_dtt <- bind_rows(kegg_list_worse_dtt) %>%
  summarize_mine(column_name = "pathway", output_column_name = "EA_term_count") %>%
  dplyr::mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  dplyr::arrange(desc(EA_term_count))
  

## For strains growing better in DTT
### Add strain tag to each dataframe before collapsing
kegg_list_better_dtt <- mapply(FUN = add_strain_tag, df = kegg_list_better_dtt, strain = names(kegg_list_better_dtt), SIMPLIFY = F)

### Collapse list of dataframes into single dataframe and add a column with the count of pathway
kegg_terms_df_better_dtt <- bind_rows(kegg_list_better_dtt) %>%
  summarize_mine(column_name = "pathway", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
```

Plot
```{r}
# For strains growing worse in DTT
## Grab only top 10 most common EA terms
temp <- kegg_terms_df_worse_dtt %>%
  dplyr::distinct(pathway, .keep_all = T) %>%
  dplyr::slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_count)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_percentage)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT") +
  ylab("Strain percentage") +
  xlab("EA term")




# For strains growing better in DTT
## Grab only top 10 most common EA terms
temp <- kegg_terms_df_better_dtt %>%
  dplyr::distinct(pathway, .keep_all = T) %>%
  dplyr::slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_count)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_percentage)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT") +
  ylab("Strain percentage") +
  xlab("EA term")
```


GO - separate GSEA results based on this classification
```{r}
# Step 1 - Separate results
go_list_better_dtt <- list()
go_list_worse_dtt <- list()

for (i in 1:length(go_gsea_results)) {
  strain_name <- names(go_gsea_results)[i]
  if (strain_name %in% strains_better_in_DTT) {
    go_list_better_dtt[[strain_name]] <- go_gsea_results[[i]]
  }
  else if (strain_name %in% strains_worse_in_DTT) {
    go_list_worse_dtt[[strain_name]] <- go_gsea_results[[i]]
  }
}



# Step 2 - Turn to dataframes
## For strains growing worse in DTT
### Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_list_worse_dtt <- mapply(FUN = add_strain_tag, df = go_list_worse_dtt, strain = names(go_list_worse_dtt), SIMPLIFY = F)

### Collapse list of dataframes into single dataframe and add a column with the count of pathway
go_terms_df_worse_dtt <- bind_rows(go_list_worse_dtt) %>%
  summarize_mine(column_name = "pathway", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
  

## For strains growing better in DTT
### Add strain tag to each dataframe before collapsing
go_list_better_dtt <- mapply(FUN = add_strain_tag, df = go_list_better_dtt, strain = names(go_list_better_dtt), SIMPLIFY = F)

### Collapse list of dataframes into single dataframe and add a column with the count of pathway
go_terms_df_better_dtt <- bind_rows(go_list_better_dtt) %>%
  summarize_mine(column_name = "pathway", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
```

Plot
```{r}
# For strains growing worse in DTT
## Grab only top 10 most common EA terms
temp <- go_terms_df_worse_dtt %>%
  dplyr::distinct(pathway, .keep_all = T) %>%
  dplyr::slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_count)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_percentage)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT") +
  ylab("Strain percentage") +
  xlab("EA term")




# For strains growing better in DTT
## Grab only top 10 most common EA terms
temp <- go_terms_df_better_dtt %>%
  dplyr::distinct(pathway, .keep_all = T) %>%
  dplyr::slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_count)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(pathway, EA_term_count), y = EA_term_percentage)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT") +
  ylab("Strain percentage") +
  xlab("EA term")
```





### 5.4.4. Produce LaTex table with significant pathways in pairs of KOs with the same anticodon
This is done in the main tRNA script. 































