---
title: "enrichment_analysis"
author: "Álvaro Gómez Pérez"
date: "2025-02-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Libraries
```{r}
library(clusterProfiler)
library(org.Sc.sgd.db)
library(data.table)
library(dplyr)
library(fgsea)
library(ggplot2)
library(jsonlite)
library(stringr)
library(forcats)
```



# 1. Load and prepare data
## 1.0. Set parameters
```{r}
# Significance level to be used for all tests and plots in this file
alpha <- 0.01

# Log-fold change limit to be considered "biologically significant"
lfc_threshold <- 1.5

# Set directories to be used
working_from = "charite"

if (working_from == "home") {
  base_dir = "/home/alvaro/MyStuff/tRNA_KOs/"
} else
if (working_from == "charite") {
  base_dir = "C:/MyStuff/tRNA_KOs/"
}
```

## 1.1. Load data
```{r}
# List of dataframes with the proteins that were DE in each KO strain - at alpha = 0.01 - remove emtpy vectors
de_proteins_list_001 <- fromJSON(paste(base_dir, "Data/Other/enrichment_analysis/de_proteins_list_001.json", sep=""))
de_proteins_list_001 <- Filter(function(x) length(x) > 0, de_proteins_list_001)

# List of dataframes with the proteins that were DE in each KO strain - at alpha = 0.05 - remove emtpy vectors
de_proteins_list_005 <- fromJSON(paste(base_dir, "Data/Other/enrichment_analysis/de_proteins_list_005.json", sep=""))
de_proteins_list_005 <- Filter(function(x) length(x) > 0, de_proteins_list_005)

# Full Entrez ID dataframe, with all genes
entrez_db <- as.data.frame(fread(paste(base_dir, "Data/Other/enrichment_analysis/entrez_reference.txt", sep="")))

# Load the dataset with that information
master_dataset <- as.data.frame(fread(paste(base_dir, "Data/Other/GtRNAdb/master_tRNA_dataset.csv", sep="")))

# Phenotypic data from Bloom-Ackermann et al., 2014
phenotypic_data <- as.data.frame(fread(paste(base_dir, "Data/Other/Articles/bloom_ackermann_2014/phenotypic_results.tsv", sep="")))


source(paste(base_dir, "Code/R/Mine/0.general_use_functions.R", sep=""))
```



# 2. GO enrichment analysis - for proteins at 0.01
## 2.1. Get NCBI gene IDs for each of the systematic name vectors I have
```{r}
de_proteins_list_entrez_ID_001 <- list()
for (i in 1:length(de_proteins_list_001)) {
  temp <- data.frame("Protein_stable_ID" = de_proteins_list_001[[i]])
  temp <- left_join(temp, entrez_db, by = "Protein_stable_ID")
  temp <- temp$`NCBI_gene_(formerly_Entrezgene)_ID`
  de_proteins_list_entrez_ID_001[[i]] <- temp
}
names(de_proteins_list_entrez_ID_001) <- names(de_proteins_list_001)
```

## 2.2. Get the information of how many DE proteins there are for each tRNA KO strain
```{r}
de_prots_per_strain <- c()
for (i in 1:length(de_proteins_list_001)) {
  temp <- de_proteins_list_001[[i]]
  de_prots_per_strain <- c(de_prots_per_strain, length(temp))
}

de_prots_per_strain_001 <- data.frame(Strain.Name = names(de_proteins_list_001),
                                      nDEP = de_prots_per_strain) %>%
  filter(Strain.Name != "WT")
```

## 2.3. Perform GO enrichment analysis
```{r}
# Come up with background (all proteins detected)
my_universe <- as.character(de_proteins_list_entrez_ID_001[["WT"]])

# Create empty list for GO results, perform GO for each KO strain, and store them there
go_results_list_001 <- list()
for (i in 2:length(de_proteins_list_entrez_ID_001)) {
  proteins_detected <- as.character(de_proteins_list_entrez_ID_001[[i]])
  if (length(proteins_detected) > 0) {
    go_results <- enrichGO(gene = proteins_detected, 
                           OrgDb = "org.Sc.sgd.db", 
                           keyType = "ENTREZID", 
                           ont = "ALL", 
                           universe = my_universe,
                           maxGSSize = 200)
    go_results <- as.data.frame(go_results)
    if (nrow(go_results) > 0) {
      go_results_list_001[[names(de_proteins_list_entrez_ID_001)[i]]] <- go_results
    }
  }
}

# Check how much correlation there is between strains having many DEPs and strains showing some GO enrichment results
has_go_results <- c()
number_of_go_terms <- c()
for (i in 1:nrow(de_prots_per_strain_001)) {
  strain <- de_prots_per_strain_001$Strain.Name[i]
  if (strain %in% names(go_results_list_001)) {
    has_go_results <- c(has_go_results, "Yes")
    number_of_go_terms <- c(number_of_go_terms, nrow(go_results_list_001[[strain]]))
  }
  else {
    has_go_results <- c(has_go_results, "No")
    number_of_go_terms <- c(number_of_go_terms, 0)
  }
}
de_prots_per_strain_001$GO_results_bp <- has_go_results
de_prots_per_strain_001$num_of_go_terms_bp <- number_of_go_terms
```

## 2.4. Add other informational columns to this dataset that might explain the success of the GO analysis for some strains
```{r}
de_prots_per_strain_001 <- left_join(de_prots_per_strain_001, master_dataset, by = "Strain.Name")
```

## 2.5. Save the EA dataset so I don't have to run all of this again
```{r}
write_json(go_results_list_001, path = paste(base_dir, "Data/Other/enrichment_analysis/go_results_list_001.json", sep=""), pretty = T)
```




# 3. GO enrichment analysis - for proteins at 0.05
## 3.1. Get NCBI gene IDs for each of the systematic name vectors I have
```{r}
de_proteins_list_entrez_ID_005 <- list()
for (i in 1:length(de_proteins_list_005)) {
  temp <- data.frame("Protein_stable_ID" = de_proteins_list_005[[i]])
  temp <- left_join(temp, entrez_db, by = "Protein_stable_ID")
  temp <- temp$`NCBI_gene_(formerly_Entrezgene)_ID`
  de_proteins_list_entrez_ID_005[[i]] <- temp
}
names(de_proteins_list_entrez_ID_005) <- names(de_proteins_list_005)
```

## 3.2. Get the information of how many DE proteins there are for each tRNA KO strain
```{r}
de_prots_per_strain <- c()
for (i in 1:length(de_proteins_list_005)) {
  temp <- de_proteins_list_005[[i]]
  de_prots_per_strain <- c(de_prots_per_strain, length(temp))
}

de_prots_per_strain_005 <- data.frame(Strain.Name = names(de_proteins_list_005),
                                      nDEP = de_prots_per_strain) %>%
  filter(Strain.Name != "WT")
```

## 3.3. Perform GO enrichment analysis
```{r}
# Come up with background (all proteins detected)
my_universe <- as.character(de_proteins_list_entrez_ID_005[["WT"]])

# Create empty list for GO results, perform GO for each KO strain, and store them there
go_results_list_005 <- list()
for (i in 2:length(de_proteins_list_entrez_ID_005)) {
  proteins_detected <- as.character(de_proteins_list_entrez_ID_005[[i]])
  if (length(proteins_detected) > 0) {
    go_results <- enrichGO(gene = proteins_detected, 
                           OrgDb = "org.Sc.sgd.db", 
                           keyType = "ENTREZID", 
                           ont = "ALL", 
                           universe = my_universe)
    go_results <- as.data.frame(go_results)
    if (nrow(go_results) > 0) {
      go_results_list_005[[names(de_proteins_list_entrez_ID_005)[i]]] <- go_results
    }
  }
}

```

## 3.4. Add other informational columns to this dataset that might explain the success of the GO analysis for some strains
```{r}
de_prots_per_strain_005 <- left_join(de_prots_per_strain_005, master_dataset, by = "Strain.Name")
```

## 3.5. Save the EA dataset so I don't have to run all of this again
```{r}
write_json(go_results_list_005, path = paste(base_dir, "Data/Other/enrichment_analysis/go_results_list_005.json", sep=""), pretty = T)
```





# 4. Evaluate overall results of EA before going into strain comparisons
## 4.1. Load data
```{r}
go_results_list_001 <- read_json(path = paste(base_dir, "Data/Other/enrichment_analysis/go_results_list_001.json", sep=""), simplifyVector = T)
go_results_list_005 <- read_json(path = paste(base_dir, "Data/Other/enrichment_analysis/go_results_list_005.json", sep=""), simplifyVector = T)
```


## 4.2. Barplots for most common EA terms
At 0.01
```{r}
# Prepare dataframe
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_results_list_001 <- mapply(FUN = add_strain_tag, df = go_results_list_001, strain = names(go_results_list_001), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_full_001 <- bind_rows(go_results_list_001) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))


# Plot
## Grab only top 10 most common EA terms
temp <- go_terms_df_full_001 %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```

At 0.05
```{r}
# Prepare dataframe
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_results_list_005 <- mapply(FUN = add_strain_tag, df = go_results_list_005, strain = names(go_results_list_005), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_full_005 <- bind_rows(go_results_list_005) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))


# Plot
## Grab only top 10 most common EA terms
temp <- go_terms_df_full_005 %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```







# 5. Compare EA results to growth in DTT (equal or worse than WT vs. better than WT) 
## 5.1. Load data
```{r}
# Results from GO EA
go_results_list_001 <- read_json(path = paste(base_dir, "Data/Other/enrichment_analysis/go_results_list_001.json", sep=""), simplifyVector = T)
go_results_list_005 <- read_json(path = paste(base_dir, "Data/Other/enrichment_analysis/go_results_list_005.json", sep=""), simplifyVector = T)

# Load the dataset with that information
master_dataset <- as.data.frame(fread(paste(base_dir, "Data/Other/GtRNAdb/master_tRNA_dataset.csv", sep="")))

# Phenotypic data from Bloom-Ackermann et al., 2014
phenotypic_data <- as.data.frame(fread(paste(base_dir, "Data/Other/Articles/bloom_ackermann_2014/phenotypic_results.tsv", sep="")))

source(paste(base_dir, "Code/R/Mine/0.general_use_functions.R", sep=""))
```


## 5.2. Evaluate growth in DTT and classify strains to better than WT or equal/worse
```{r}
# Plot GR and GY in DTT
ggplot(data = phenotypic_data, aes(x = DTT)) +
  geom_histogram(binwidth = 1) +
  theme_light() +
  geom_vline(xintercept = mean(phenotypic_data$DTT, na.rm = T)+sd(phenotypic_data$DTT, na.rm = T), col = "red") +
  geom_vline(xintercept = mean(phenotypic_data$DTT, na.rm = T)-sd(phenotypic_data$DTT, na.rm = T), col = "red")

ggplot(data = phenotypic_data, aes(x = DTT_GY)) +
  geom_histogram(binwidth = 1) +
  theme_light() +
  geom_vline(xintercept = mean(phenotypic_data$DTT_GY, na.rm = T)+sd(phenotypic_data$DTT_GY, na.rm = T), col = "red") +
  geom_vline(xintercept = mean(phenotypic_data$DTT_GY, na.rm = T)-sd(phenotypic_data$DTT_GY, na.rm = T), col = "red")

# At first I just divided strains into those above and below 0, and grabbed those that were above 0 in both GR and GY as one group, and those that were below 0 in both GR and GY as the other group. However, I found no pattern at all, so I think it would make sense to be a bit more stringent in how I select the strains. I am going to do it based on the SD of each of the distribution above

# Need to look into the original article to figure out which one to use - going with both for now, so I don't take all strains here, only those that agree on their GR and GY (not sure if that makes sense, going by vibes for now)
## Need to get rid of the lethal strains before I select here
phenotypic_data <- phenotypic_data %>%
  filter(lethal != "Yes",
         !(is.na(DTT)),
         !is.na(DTT_GY))

## Actually perform the selection
strains_better_in_DTT <- phenotypic_data$gene_name[phenotypic_data$DTT > sd(phenotypic_data$DTT, na.rm = T) & phenotypic_data$DTT_GY > sd(phenotypic_data$DTT_GY, na.rm = T)]
strains_worse_or_equal_in_DTT <- phenotypic_data$gene_name[phenotypic_data$DTT < -sd(phenotypic_data$DTT, na.rm = T) & phenotypic_data$DTT_GY < -sd(phenotypic_data$DTT_GY, na.rm = T)]
```

## 5.2. Compare EA results to these strain classification
```{r}
go_list_better_dtt <- list()
go_list_worse_dtt <- list()

for (i in 1:length(go_results_list_001)) {
  strain_name <- names(go_results_list_001)[i]
  if (strain_name %in% strains_better_in_DTT) {
    go_list_better_dtt[[strain_name]] <- go_results_list_001[[i]]
  }
  else if (strain_name %in% strains_worse_or_equal_in_DTT) {
    go_list_worse_dtt[[strain_name]] <- go_results_list_001[[i]]
  }
}
```

## 5.3. Come up with a count of the enriched pathways throughout each set of strains
Get dataframes
```{r}
# For strains growing worse in DTT
## Add strain tag to each dataframe before collapsing
add_strain_tag <- function(df, strain) {
  df$Strain <- rep(strain, nrow(df))
  return(df)
}
go_list_worse_dtt <- mapply(FUN = add_strain_tag, df = go_list_worse_dtt, strain = names(go_list_worse_dtt), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_worse_dtt <- bind_rows(go_list_worse_dtt) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
  

# For strains growing better in DTT
## Add strain tag to each dataframe before collapsing
go_list_better_dtt <- mapply(FUN = add_strain_tag, df = go_list_better_dtt, strain = names(go_list_better_dtt), SIMPLIFY = F)

## Collapse list of dataframes into single dataframe and add a column with the count of Description
go_terms_df_better_dtt <- bind_rows(go_list_better_dtt) %>%
  summarize_mine(column_name = "Description", output_column_name = "EA_term_count") %>%
  mutate(EA_term_percentage = EA_term_count/length(unique(Strain))) %>%
  arrange(desc(EA_term_count))
```

Plot
```{r}
# For strains growing worse in DTT
## Grab only top 10 most common EA terms
temp <- go_terms_df_worse_dtt %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")




# For strains growing better in DTT
## Grab only top 10 most common EA terms
temp <- go_terms_df_better_dtt %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Across strains growing better than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```

--> Pretty much the same in both subsets, let's see what is actually different between them
```{r}
# Filter the dataframes
ea_only_worse_df <- go_terms_df_worse_dtt[!(go_terms_df_worse_dtt$Description %in% go_terms_df_better_dtt$Description),]
ea_only_better <- go_terms_df_better_dtt[!(go_terms_df_better_dtt$Description %in% go_terms_df_worse_dtt$Description),]



# Plot
# For strains growing worse in DTT
## Grab only top 10 most common EA terms
temp <- ea_only_worse_df %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")




# For strains growing better in DTT
## Grab only top 10 most common EA terms
temp <- ea_only_better %>%
  distinct(Description, .keep_all = T) %>%
  slice(1:10)

## Plot with total EA term count
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_count, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing better than WT in DTT",
       fill = "Ontology") +
  ylab("Strain count") +
  xlab("EA term")

# Plot with % - total EA term count/number of strains
ggplot(data = temp, aes(x = fct_reorder(Description, EA_term_count), y = EA_term_percentage, fill = ONTOLOGY)) +
  geom_col() +
  theme_light() +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1)) +
  labs(title = "Most common EA terms",
       subtitle = "Unique to strains growing worse than WT in DTT",
       fill = "Ontology") +
  ylab("Strain percentage") +
  xlab("EA term")
```










# 6. GSEA - try to imitate the results from the original article
## 6.1. Prepare ranked gene lists
```{r}
# Load list of all dataframes (one per strain) wiht the DE results
da <- read_json(paste(base_dir, "Data/Other/enrichment_analysis/da.json", sep=""), simplifyVector = T)
da <- da[names(da) != "WT"]

# Create empty list to put the named ranked gene lists in
ranked_genes <- list()

# Iterate over these dataframes
for (i in 1:length(da)) {
  temp <- da[[i]]
  
  # Using as score the -log10(p-value)*sign of the logFC
  rankings <- sign(temp$logFC)*(-log10(temp$adj.P.Val))
  names(rankings) <- rownames(temp)
  rankings <- sort(rankings, decreasing = TRUE)
  
  # Plot
  plot(rankings)
  
  # Check max and min - to see if there is a need to correct for -Inf and +Inf
  max(rankings)
  min(rankings)
  
  # Some genes have such low p values that the signed pval is +- inf, we need to change it to the maximum * constant to avoid problems with fgsea
  # (this part is copy pasted from the link)
  max_ranking <- max(rankings[is.finite(rankings)])
  min_ranking <- min(rankings[is.finite(rankings)])
  rankings <- replace(rankings, rankings > max_ranking, max_ranking * 10)
  rankings <- replace(rankings, rankings < min_ranking, min_ranking * 10)
  rankings <- sort(rankings, decreasing = TRUE) # sort genes by ranking
  
  # Add them to the list where we are saving them
  ranked_genes[[names(da)[i]]] <- rankings
}
```

## 6.2. Perform GSEA with KEGG annotation
### Download KEGG gene lists (according to https://www.researchgate.net/post/How_i_can_get_a_list_of_KEGG_pathways_and_its_list_of_genes)
DO NOT RUN AGAIN - I used this to download the file, I just need to load it, but I left this code here in case I want to know how I did it
```{r}
BiocManager::install("KEGGREST")
BiocManager::install("EnrichmentBrowser")
library("KEGGREST")
library("EnrichmentBrowser")

#step2: check and obtain a list of entry identifiers (in this case: sce) and associated definition for a given database or a given set of database entries.
MRSA252 <- keggList("sce")

#step 3: download the pathways of that organism:
sarpathway <- downloadPathways("sce")

#step 4: retrieve gene sets for an organism from databases such as GO and KEGG:
sar <- getGenesets(org = "sce", db = "kegg", cache = TRUE, return.type="list")

#step5: Parse and write the gene sets to a flat text file in GMT format for other pathway enrichment analysis programs (e.g., GSEA):
writeGMT(sar, gmt.file = paste(base_dir, "Data/Other/enrichment_analysis/20250305_kegg_sce_gmt", sep=""))
```


### Now perform the GSEA as described by https://biostatsquid.com/fgsea-tutorial-gsea/
Get a list of pathways, with each entry being a vector with the Entrez IDs for the genes involved in that pathway. Make it so that only proteins that we detect in our dataset are included here. 
```{r}
# Read in the .gmt file
gmt <- read.gmt(paste(base_dir, "Data/Other/enrichment_analysis/20250305_kegg_sce_gmt", sep=""))
gmt$term <- as.character(gmt$term)

# Get a simple vector with the Entrez IDs of the proteins we detect
trna_ko <- as.data.frame(as.matrix(fread(paste(base_dir, "Data/Other/proteomics_data/processed_proteomics_dataframe.tsv", sep="")), rownames = 1))
yeastmine <- as.data.frame(fread(paste(base_dir, "Data/alliancemine_results_2024-11-20T11-19-04.tsv", sep="")))
proteins_detected <- match_systematic_and_standard_protein_names(data = rownames(trna_ko),
                                                                 yeastmine = yeastmine,
                                                                 input = "standard",
                                                                 simplify = T)
proteins_detected <- entrez_db$`NCBI_gene_(formerly_Entrezgene)_ID`[entrez_db$Protein_stable_ID %in% proteins_detected]

# Get our list of vectors, where each vector is a KEGG pathway and the terms in it are the genes related to it
gmt <- gmt %>%
  mutate(pathway_code = substr(term, 1, str_locate(term, "_")[1]-1), 
         pathway_name = substr(term, str_locate(term, "_")[1]+1, nchar(term)))

kegg_genes <- list()
pathways <- unique(gmt$pathway_name)
for (i in 1:length(pathways)) {
  pathway <- pathways[i]
  temp <- gmt %>%
    filter(pathway_name == pathway)
  kegg_genes[[pathway]] <- as.character(temp$gene)[as.character(temp$gene) %in% proteins_detected]
}
```

Run GSEA
The warning I get here refers to many proteins having exactly the same rank (these are the ones we see in a horizontal line in the plots produced in 5.1.) - this should be okay, since these are the genes that were not significantly differentially expressed at all - but tbh this makes it so that I cannot really get that much info from those anyway, right? So there isn't that much of a difference between this and doing the GSEA only with DE proteins, right?? Idk
```{r}
kegg_gsea_results <- list()

for (i in 1:length(ranked_genes)) {
  kegg_gsea_results[[names(ranked_genes)[i]]] <- fgsea(pathways = kegg_genes, # List of gene sets to check
                 stats = ranked_genes[[i]],
                 scoreType = 'std', # in this case we have both pos and neg rankings. if only pos or neg, set to 'pos', 'neg'
                 minSize = 1,
                 maxSize = 1000,
                 nproc = 1) # for parallelisation
}
```

THE RESULTS JUST COME OUT TOTALLY EMPTY FOR ALL KOS :D - I GUESS I'LL HAVE TO LOOK INTO THIS AT SOME POINT





## 6.3. Perform GSEA with Reactome annotation??














## 2.5. THIS IS WHAT I HAD WRITTEN ABOVE BEFORE (WITHIN GO ENRICHMENT ANALYSIS) BEFORE REALIZING I NEEDED TO DO GSEA SEPARATELY, JUST IN CASE I CAN REUSE THE TEXT, EVEN IF NOT THE CODE

### For SC strains: pathways related to proteotoxic stress were significantly induced w.r.t. to the MC group!! That is not what I am checking here, keep that in mind. These include:
  - Proteasome
  - Protein processing in endoplasmic reticulum
```{r}
SC_strains <- phenotypic_data$gene_name[phenotypic_data$MC_or_SC == "SC"]

for (i in 1:length(go_results_list_001)) {
  if (names(go_results_list_001)[i] %in% SC_strains) {
    print(go_results_list_001[[i]][["Description"]])
  }
}
```

### For MC strains: pathways related to translation were significantly induced w.r.t. to the SC group!! That is not what I am checking here, keep that in mind. These include:
  - Ribosome biogenesis
  - Ribosome
```{r}
MC_strains <- phenotypic_data$gene_name[phenotypic_data$MC_or_SC == "MC"]

for (i in 1:length(go_results_list_001)) {
  if (names(go_results_list_001)[i] %in% MC_strains) {
    print(go_results_list_001[[i]][["Description"]])
  }
}
```


































